// texas-poker-server/src/websocket/handlers/gameHandler.ts
import { Socket, Server } from 'socket.io';
import { RoomManager } from '../../room/RoomManager';
import { GameEngine } from '../../game/GameEngine';
import { ClientEvents, ServerEvents } from '../../types/events';
import { PlayerAction, GamePhase } from '../../types/poker';
import { RoomStatus } from '../../types/room';
import { addActionLog, loadRoomLogs } from '../../room/ActionLogManager';

export const gameEngines: Map<string, GameEngine> = new Map();

function safeCallback(callback: any, response: any): void {
  if (typeof callback === 'function') {
    callback(response);
  }
}

export function handleGameEvents(socket: Socket, io: Server, roomManager: RoomManager): void {
  socket.on(ClientEvents.PLAYER_ACTION, (data: { action: string; amount?: number }, callback?: (response: any) => void) => {
    try {
      const playerId = socket.data.playerId;
      if (!playerId) {
        safeCallback(callback, { success: false, error: '未登录' });
        return;
      }

      const roomId = roomManager.getPlayerRoomId(playerId);
      if (!roomId) {
        safeCallback(callback, { success: false, error: '你不在任何房间中' });
        return;
      }

      const room = roomManager.getRoom(roomId);
      if (!room) {
        safeCallback(callback, { success: false, error: '房间不存在' });
        return;
      }

      let gameEngine = gameEngines.get(roomId);
      if (!gameEngine) {
        safeCallback(callback, { success: false, error: '游戏引擎未找到' });
        return;
      }

      const actionMap: Record<string, PlayerAction> = {
        'fold': PlayerAction.FOLD,
        'check': PlayerAction.CHECK,
        'call': PlayerAction.CALL,
        'raise': PlayerAction.RAISE,
        'all-in': PlayerAction.ALL_IN,
        'allin': PlayerAction.ALL_IN,
      };

      const playerAction = actionMap[data.action.toLowerCase()];
      if (!playerAction) {
        safeCallback(callback, { success: false, error: `无效操作: ${data.action}` });
        return;
      }

      const result = gameEngine.performAction(playerId, playerAction, data.amount);

      if (result.success) {
        const gameState = gameEngine.getState();
        room.gameState = gameState;

        syncPlayerChipsToRoom(gameEngine, room);

        const actor = room.players.find((p: any) => p.id === playerId);
        if (actor) {
          loadRoomLogs(roomId);
          addActionLog(roomId, gameState.handId || '', playerId, actor.name, data.action, data.amount, gameState.phase);
        }

        const isGameEnding = gameState.phase === GamePhase.SHOWDOWN || gameState.phase === GamePhase.ENDED;

        io.to(roomId).emit(ServerEvents.ACTION_RESULT, {
          playerId,
          playerName: actor?.name || playerId,
          action: data.action,
          amount: data.amount,
          gameState: sanitizeGameState(gameState),
          ...(isGameEnding ? {} : { room: sanitizeRoom(room) }),
        });

        if (isGameEnding) {
          const { winners, potResults, allHands } = gameEngine.showdown();

          const finalGameState = gameEngine.getState();
          room.gameState = finalGameState;

          syncPlayerChipsToRoom(gameEngine, room);

          autoRebuyBustedPlayers(room, io);

          for (const w of winners) {
            const roomPlayer = room.players.find((p: any) => p.id === w.playerId);
            if (roomPlayer) {
              w.playerName = roomPlayer.name;
            }
          }
          for (const h of allHands) {
            const roomPlayer = room.players.find((p: any) => p.id === h.playerId);
            if (roomPlayer) {
              h.playerName = roomPlayer.name;
            }
          }

          const mergedWinners = (() => {
            const map = new Map<string, any>();
            for (const w of winners) {
              const existing = map.get(w.playerId);
              if (existing) {
                existing.winAmount += w.winAmount;
                if (w.potType === 'side') {
                  existing.potType = 'both';
                }
              } else {
                map.set(w.playerId, { ...w });
              }
            }
            for (const [, mw] of map) {
              const allHand = allHands.find((h: any) => h.playerId === mw.playerId && h.isWinner);
              if (allHand && allHand.winAmount !== undefined) {
                mw.winAmount = allHand.winAmount;
              }
            }
            return Array.from(map.values()).filter((mw: any) => {
              const allHand = allHands.find((h: any) => h.playerId === mw.playerId);
              return allHand && allHand.isWinner;
            });
          })();

          room.status = RoomStatus.WAITING;

          const currentGamePlayers = gameEngine.getPlayers();
          const currentGamePlayerIds = new Set(currentGamePlayers.map(p => p.id));
          for (const p of room.players) {
            if (currentGamePlayerIds.has(p.id)) {
              p.isReady = false;
            }
          }

          io.to(roomId).emit(ServerEvents.SHOWDOWN, {
            winners: mergedWinners,
            potResults,
            allHands,
            communityCards: finalGameState.communityCards,
            gameState: sanitizeGameState(finalGameState),
            room: sanitizeRoom(room),
          });

          io.to(roomId).emit(ServerEvents.HAND_RESULT, {
            winners: mergedWinners,
            potResults,
            allHands,
            communityCards: finalGameState.communityCards,
            room: sanitizeRoom(room),
          });

          room.gameState = {
            ...room.gameState,
            currentBet: 0,
            minRaise: room.config?.bigBlind || 20,
            roundBets: {},
            pots: [],
            totalPot: 0,
            actions: [],
            communityCards: [],
            playerCards: {},
            playerStatus: {},
            playerRoles: {},
            lastRaiseIndex: -1,
            currentPlayerIndex: -1,
            currentPlayerId: '',
          };

          io.to(roomId).emit(ServerEvents.ROOM_UPDATED, {
            type: 'updated',
            room: sanitizeRoom(room),
          });
        } else {
          const currentPlayerId = gameEngine.getCurrentPlayerId();
          if (currentPlayerId) {
            const currentPlayer = room.players.find(p => p.id === currentPlayerId);
            if (currentPlayer) {
              io.to(roomId).emit(ServerEvents.PLAYER_TURN, {
                playerId: currentPlayerId,
                playerName: currentPlayer.name,
                timeout: 30,
                validActions: gameEngine.getValidActions(currentPlayerId),
              });
            } else {
              io.to(roomId).emit(ServerEvents.PLAYER_TURN, {
                playerId: currentPlayerId,
                playerName: currentPlayerId,
                timeout: 30,
                validActions: gameEngine.getValidActions(currentPlayerId),
              });
            }
          } else {
            const playingPlayers = room.players.filter((p: any) =>
              gameState.playerStatus?.[p.id] === 'playing'
            );
            if (playingPlayers.length > 0) {
              const firstActive = playingPlayers[0];
              io.to(roomId).emit(ServerEvents.PLAYER_TURN, {
                playerId: firstActive.id,
                playerName: firstActive.name,
                timeout: 30,
                validActions: gameEngine.getValidActions(firstActive.id),
              });
            }
          }
        }

        safeCallback(callback, { success: true });
      } else {
        safeCallback(callback, { success: false, error: result.error });
      }
    } catch (error) {
      safeCallback(callback, { success: false, error: '执行动作失败' });
    }
  });

  socket.on(ClientEvents.SEND_CHAT, (data: { message: string }, callback?: (response: any) => void) => {
    try {
      const playerId = socket.data.playerId;
      if (!playerId) {
        safeCallback(callback, { success: false, error: '未登录' });
        return;
      }

      const roomId = roomManager.getPlayerRoomId(playerId);
      if (!roomId) {
        safeCallback(callback, { success: false, error: '你不在任何房间中' });
        return;
      }

      const room = roomManager.getRoom(roomId);
      const player = room?.players.find(p => p.id === playerId);

      if (player) {
        io.to(roomId).emit(ServerEvents.CHAT_MESSAGE, {
          playerId,
          playerName: player.name,
          message: data.message,
          timestamp: Date.now(),
        });
      }

      safeCallback(callback, { success: true });
    } catch (error) {
      safeCallback(callback, { success: false, error: '发送消息失败' });
    }
  });
}

function syncPlayerChipsToRoom(gameEngine: GameEngine, room: any): void {
  const enginePlayers = gameEngine.getPlayers();
  for (const ep of enginePlayers) {
    const roomPlayer = room.players.find((p: any) => p.id === ep.id);
    if (roomPlayer) {
      roomPlayer.chips = ep.chips;
    }
  }
}

function getAllHandsForShowdown(gameEngine: GameEngine, players: any[]): any[] {
  const state = gameEngine.getState();
  const allHands = [];

  for (const player of players) {
    const cards = gameEngine.getPlayerCards(player.id);
    if (cards && state.playerStatus[player.id] !== 'folded') {
      allHands.push({
        playerId: player.id,
        playerName: player.name,
        holeCards: cards,
      });
    }
  }

  return allHands;
}

function sanitizeGameState(gameState: any): any {
  const sanitized = JSON.parse(JSON.stringify(gameState));
  sanitized.playerCards = {};
  return sanitized;
}

function sanitizeRoom(room: any): any {
  return {
    config: room.config,
    status: room.status,
    players: room.players.map((p: any) => ({
      id: p.id,
      name: p.name,
      avatar: p.avatar,
      seatIndex: p.seatIndex,
      chips: p.chips,
      totalBuyIn: p.totalBuyIn,
      isReady: p.isReady,
      isOnline: p.isOnline,
    })),
  };
}

function autoRebuyBustedPlayers(room: any, io: any): void {
  const buyInAmount = room.config.buyInMin;
  for (const player of room.players) {
    if (player.chips <= 0) {
      player.chips = buyInAmount;
      player.totalBuyIn += buyInAmount;
      io.to(room.config.roomId).emit(ServerEvents.CHIPS_RECEIVED, {
        playerId: player.id,
        amount: buyInAmount,
        autoRebuy: true,
        room: sanitizeRoom(room),
      });
    }
  }
}

export function setGameEngine(roomId: string, gameEngine: GameEngine): void {
  gameEngines.set(roomId, gameEngine);
}
