// texas-poker-server/dist/game/GameEngine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameEngine = void 0;
const uuid_1 = require("uuid");
const poker_1 = require("../types/poker");
const Deck_1 = require("../poker/Deck");
const HandEvaluator_1 = require("../poker/HandEvaluator");
class GameEngine {
    constructor(players, dealerIndex, config) {
        this.hasActedThisRound = new Set();
        this.lastAggressorIndex = -1;
        this.actionCount = 0;
        this.playerInitialChips = new Map();
        this.players = players.filter(p => p.isReady && p.chips > 0);
        this.config = config;
        this.deck = new Deck_1.Deck();
        const playerIds = this.players.map(p => p.id);
        this.state = {
            handId: (0, uuid_1.v4)(),
            phase: poker_1.GamePhase.WAITING,
            deck: [],
            communityCards: [],
            pots: [],
            totalPot: 0,
            currentPlayerIndex: -1,
            currentPlayerId: '',
            dealerIndex,
            smallBlindIndex: (dealerIndex + 1) % this.players.length,
            bigBlindIndex: (dealerIndex + 2) % this.players.length,
            lastRaiseIndex: -1,
            currentBet: 0,
            minRaise: config.bigBlind,
            roundBets: {},
            playerCards: {},
            playerStatus: Object.fromEntries(playerIds.map(id => [id, poker_1.PlayerStatus.WAITING])),
            playerRoles: {},
            actions: [],
            startTime: Date.now(),
        };
    }
    start() {
        if (this.players.length < 2) {
            throw new Error('至少需要2名玩家才能开始游戏');
        }
        this.deck = new Deck_1.Deck();
        this.deck.shuffle();
        this.state.communityCards = [];
        this.state.pots = [];
        this.state.totalPot = 0;
        this.state.roundBets = {};
        this.state.currentBet = 0;
        this.state.lastRaiseIndex = -1;
        this.state.actions = [];
        this.hasActedThisRound = new Set();
        this.lastAggressorIndex = -1;
        this.actionCount = 0;
        for (const player of this.players) {
            this.state.playerStatus[player.id] = poker_1.PlayerStatus.PLAYING;
            this.state.roundBets[player.id] = 0;
            this.playerInitialChips.set(player.id, player.chips);
        }
        this.state.phase = poker_1.GamePhase.PRE_FLOP;
        this.initPlayerRoles();
        this.dealHoleCards();
        this.postBlinds();
        this.state.currentPlayerIndex = this.getNextActivePlayerIndex(this.state.bigBlindIndex);
        this.state.currentPlayerId = this.players[this.state.currentPlayerIndex]?.id || '';
        return this.state;
    }
    initPlayerRoles() {
        this.state.playerRoles = {};
        const n = this.players.length;
        const dealer = this.players[this.state.dealerIndex % n];
        const sb = this.players[this.state.smallBlindIndex % n];
        const bb = this.players[this.state.bigBlindIndex % n];
        if (dealer)
            this.state.playerRoles[dealer.id] = poker_1.PlayerRole.DEALER;
        if (sb)
            this.state.playerRoles[sb.id] = poker_1.PlayerRole.SB;
        if (bb)
            this.state.playerRoles[bb.id] = poker_1.PlayerRole.BB;
    }
    dealHoleCards() {
        for (const player of this.players) {
            this.state.playerCards[player.id] = [
                this.deck.deal(),
                this.deck.deal(),
            ];
        }
    }
    postBlinds() {
        const n = this.players.length;
        const sb = this.players[this.state.smallBlindIndex % n];
        const bb = this.players[this.state.bigBlindIndex % n];
        if (sb) {
            const sbAmount = Math.min(this.config.smallBlind, sb.chips);
            sb.chips -= sbAmount;
            this.state.roundBets[sb.id] = sbAmount;
            if (sb.chips === 0) {
                this.state.playerStatus[sb.id] = poker_1.PlayerStatus.ALL_IN;
            }
        }
        if (bb) {
            const bbAmount = Math.min(this.config.bigBlind, bb.chips);
            bb.chips -= bbAmount;
            this.state.roundBets[bb.id] = bbAmount;
            this.state.currentBet = bbAmount;
            if (bb.chips === 0) {
                this.state.playerStatus[bb.id] = poker_1.PlayerStatus.ALL_IN;
            }
        }
        this.state.minRaise = this.config.bigBlind;
        this.lastAggressorIndex = this.state.bigBlindIndex % n;
        this.state.totalPot = this.calcTotalPot();
    }
    getNextActivePlayerIndex(fromIndex) {
        const n = this.players.length;
        let index = (fromIndex + 1) % n;
        let checks = 0;
        while (checks < n) {
            const player = this.players[index];
            const status = this.state.playerStatus[player.id];
            if (status === poker_1.PlayerStatus.PLAYING) {
                return index;
            }
            index = (index + 1) % n;
            checks++;
        }
        return -1;
    }
    performAction(playerId, action, amount) {
        const playerIndex = this.players.findIndex(p => p.id === playerId);
        if (playerIndex === -1) {
            return { success: false, error: '玩家不存在' };
        }
        const player = this.players[playerIndex];
        if (this.state.playerStatus[player.id] !== poker_1.PlayerStatus.PLAYING) {
            return { success: false, error: '玩家不在游戏中' };
        }
        if (playerIndex !== this.state.currentPlayerIndex) {
            return { success: false, error: '不是你的回合' };
        }
        const myBet = this.state.roundBets[player.id] || 0;
        const toCall = this.state.currentBet - myBet;
        switch (action) {
            case poker_1.PlayerAction.FOLD:
                this.state.playerStatus[player.id] = poker_1.PlayerStatus.FOLDED;
                break;
            case poker_1.PlayerAction.CHECK:
                if (toCall > 0) {
                    return { success: false, error: '前面有人下注，不能过牌' };
                }
                break;
            case poker_1.PlayerAction.CALL: {
                if (toCall <= 0) {
                    return { success: false, error: '没有需要跟注的金额' };
                }
                const callAmount = Math.min(toCall, player.chips);
                player.chips -= callAmount;
                this.state.roundBets[player.id] = myBet + callAmount;
                if (player.chips === 0) {
                    this.state.playerStatus[player.id] = poker_1.PlayerStatus.ALL_IN;
                }
                break;
            }
            case poker_1.PlayerAction.RAISE: {
                if (!amount || amount <= 0) {
                    return { success: false, error: '加注金额必须大于0' };
                }
                const totalRaiseAmount = amount;
                const callPlusRaise = toCall + totalRaiseAmount;
                const actualAmount = Math.min(callPlusRaise, player.chips);
                player.chips -= actualAmount;
                this.state.roundBets[player.id] = myBet + actualAmount;
                this.state.currentBet = myBet + actualAmount;
                this.state.minRaise = totalRaiseAmount;
                this.lastAggressorIndex = playerIndex;
                this.hasActedThisRound.add(player.id);
                if (player.chips === 0) {
                    this.state.playerStatus[player.id] = poker_1.PlayerStatus.ALL_IN;
                }
                break;
            }
            case poker_1.PlayerAction.ALL_IN: {
                const allInAmount = player.chips;
                player.chips = 0;
                this.state.roundBets[player.id] = myBet + allInAmount;
                if (myBet + allInAmount > this.state.currentBet) {
                    this.state.currentBet = myBet + allInAmount;
                    this.lastAggressorIndex = playerIndex;
                }
                this.state.playerStatus[player.id] = poker_1.PlayerStatus.ALL_IN;
                this.hasActedThisRound.add(player.id);
                break;
            }
            default:
                return { success: false, error: '无效的动作' };
        }
        if (action !== poker_1.PlayerAction.RAISE && action !== poker_1.PlayerAction.ALL_IN) {
            this.hasActedThisRound.add(player.id);
        }
        this.actionCount++;
        this.state.actions.push({
            playerId,
            playerName: player.name,
            action: action.toString(),
            amount: amount || 0,
            timestamp: Date.now(),
            phase: this.state.phase,
        });
        this.state.totalPot = this.calcTotalPot();
        if (this.checkOnlyOnePlayerLeft()) {
            this.endHand();
        }
        else if (this.isBettingRoundComplete()) {
            this.advancePhase();
        }
        else {
            this.advanceToNextPlayer();
        }
        return { success: true };
    }
    checkOnlyOnePlayerLeft() {
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.PLAYING);
        const allInPlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.ALL_IN);
        return activePlayers.length + allInPlayers.length <= 1;
    }
    isBettingRoundComplete() {
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.PLAYING);
        if (activePlayers.length === 0) {
            return true;
        }
        for (const player of activePlayers) {
            const myBet = this.state.roundBets[player.id] || 0;
            if (myBet < this.state.currentBet) {
                return false;
            }
            if (!this.hasActedThisRound.has(player.id)) {
                return false;
            }
        }
        return true;
    }
    advanceToNextPlayer() {
        const nextIndex = this.getNextActivePlayerIndex(this.state.currentPlayerIndex);
        if (nextIndex === -1) {
            this.state.currentPlayerIndex = -1;
            this.state.currentPlayerId = '';
            this.advancePhase();
            return;
        }
        this.state.currentPlayerIndex = nextIndex;
        this.state.currentPlayerId = this.players[nextIndex]?.id || '';
    }
    advancePhase() {
        this.collectBets();
        this.hasActedThisRound = new Set();
        this.lastAggressorIndex = -1;
        this.state.currentBet = 0;
        this.state.roundBets = {};
        this.actionCount = 0;
        this.state.totalPot = this.calcTotalPot();
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.PLAYING ||
            this.state.playerStatus[p.id] === poker_1.PlayerStatus.ALL_IN);
        if (activePlayers.length <= 1) {
            this.endHand();
            return;
        }
        const playingPlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.PLAYING);
        switch (this.state.phase) {
            case poker_1.GamePhase.PRE_FLOP:
                this.state.phase = poker_1.GamePhase.FLOP;
                this.deck.deal();
                this.state.communityCards.push(this.deck.deal(), this.deck.deal(), this.deck.deal());
                break;
            case poker_1.GamePhase.FLOP:
                this.state.phase = poker_1.GamePhase.TURN;
                this.deck.deal();
                this.state.communityCards.push(this.deck.deal());
                break;
            case poker_1.GamePhase.TURN:
                this.state.phase = poker_1.GamePhase.RIVER;
                this.deck.deal();
                this.state.communityCards.push(this.deck.deal());
                break;
            case poker_1.GamePhase.RIVER:
                this.endHand();
                return;
            default:
                return;
        }
        if (playingPlayers.length <= 1) {
            this.dealRemainingCommunityCards();
            this.endHand();
            return;
        }
        const nextIndex = this.getNextActivePlayerIndex(this.state.dealerIndex);
        if (nextIndex === -1) {
            this.dealRemainingCommunityCards();
            this.endHand();
            return;
        }
        this.state.currentPlayerIndex = nextIndex;
        this.state.currentPlayerId = this.players[nextIndex]?.id || '';
    }
    dealRemainingCommunityCards() {
        const needed = 5 - this.state.communityCards.length;
        if (needed <= 0)
            return;
        if (this.state.communityCards.length === 0) {
            this.deck.deal();
            this.state.communityCards.push(this.deck.deal(), this.deck.deal(), this.deck.deal());
        }
        if (this.state.communityCards.length < 5) {
            this.deck.deal();
            this.state.communityCards.push(this.deck.deal());
        }
        if (this.state.communityCards.length < 5) {
            this.deck.deal();
            this.state.communityCards.push(this.deck.deal());
        }
        this.state.phase = poker_1.GamePhase.RIVER;
    }
    collectBets() {
        const allBets = [];
        for (const player of this.players) {
            const bet = this.state.roundBets[player.id] || 0;
            if (bet > 0) {
                allBets.push({
                    playerId: player.id,
                    bet,
                    folded: this.state.playerStatus[player.id] === poker_1.PlayerStatus.FOLDED,
                });
            }
        }
        if (allBets.length === 0)
            return;
        const uniqueLevels = [...new Set(allBets.map(b => b.bet))].sort((a, b) => a - b);
        let prevLevel = 0;
        for (const level of uniqueLevels) {
            const levelDiff = level - prevLevel;
            let potAmount = 0;
            const eligibleIds = [];
            for (const entry of allBets) {
                if (entry.bet >= level) {
                    potAmount += levelDiff;
                    if (!entry.folded) {
                        eligibleIds.push(entry.playerId);
                    }
                }
            }
            if (potAmount > 0 && eligibleIds.length > 0) {
                this.state.pots.push({
                    id: (0, uuid_1.v4)(),
                    amount: potAmount,
                    eligiblePlayers: eligibleIds,
                });
            }
            prevLevel = level;
        }
    }
    endHand() {
        this.collectBets();
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] !== poker_1.PlayerStatus.FOLDED);
        if (activePlayers.length === 0) {
            this.state.phase = poker_1.GamePhase.SHOWDOWN;
            return;
        }
        if (activePlayers.length === 1) {
            const winner = activePlayers[0];
            const totalPot = this.state.pots.reduce((sum, p) => sum + p.amount, 0);
            winner.chips += totalPot;
            this.state.phase = poker_1.GamePhase.SHOWDOWN;
            return;
        }
        this.determineWinners();
        this.state.phase = poker_1.GamePhase.SHOWDOWN;
    }
    determineWinners() {
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] !== poker_1.PlayerStatus.FOLDED);
        if (activePlayers.length === 0)
            return;
        const playerHands = new Map();
        for (const player of activePlayers) {
            const holeCards = this.state.playerCards[player.id];
            const communityCards = this.state.communityCards;
            if (holeCards && communityCards.length >= 3) {
                const allCards = [...holeCards, ...communityCards];
                const hand = HandEvaluator_1.HandEvaluator.evaluate(allCards);
                playerHands.set(player.id, { hand, cards: allCards });
            }
        }
        for (const pot of this.state.pots) {
            const eligiblePlayers = activePlayers.filter(p => pot.eligiblePlayers.includes(p.id));
            if (eligiblePlayers.length === 0)
                continue;
            let bestHand = null;
            let potWinnerIds = [];
            for (const player of eligiblePlayers) {
                const handData = playerHands.get(player.id);
                if (!handData)
                    continue;
                const hand = handData.hand;
                if (!bestHand || hand.rank > bestHand.rank ||
                    (hand.rank === bestHand.rank && hand.value > bestHand.value)) {
                    bestHand = hand;
                    potWinnerIds = [player.id];
                }
                else if (hand.rank === bestHand.rank && hand.value === bestHand.value) {
                    potWinnerIds.push(player.id);
                }
            }
            const splitAmount = Math.floor(pot.amount / potWinnerIds.length);
            const remainder = pot.amount - splitAmount * potWinnerIds.length;
            for (let i = 0; i < potWinnerIds.length; i++) {
                const winner = this.players.find(p => p.id === potWinnerIds[i]);
                if (winner) {
                    winner.chips += splitAmount + (i === 0 ? remainder : 0);
                }
            }
        }
    }
    showdown() {
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] !== poker_1.PlayerStatus.FOLDED);
        const winners = [];
        const potResults = [];
        const allHands = [];
        if (activePlayers.length === 0) {
            this.state.phase = poker_1.GamePhase.ENDED;
            return { winners, potResults, allHands };
        }
        if (activePlayers.length === 1) {
            const winner = activePlayers[0];
            const totalPot = this.state.pots.reduce((sum, p) => sum + p.amount, 0);
            const initialChips = this.playerInitialChips.get(winner.id) || 0;
            const netWin = winner.chips - initialChips;
            winners.push({
                playerId: winner.id,
                playerName: winner.name,
                winAmount: netWin,
                potType: 'main',
                handRank: 'win',
                handDescription: '其他玩家弃牌',
                winningCards: this.state.playerCards[winner.id] || [],
                holeCards: this.state.playerCards[winner.id] || [],
                explanation: `${winner.name}获胜，其他玩家弃牌`,
            });
            potResults.push({
                potId: 'pot-0',
                amount: totalPot,
                winners: [winner.id],
                splitAmount: totalPot,
                remainder: 0,
            });
            allHands.push({
                playerId: winner.id,
                playerName: winner.name,
                holeCards: this.state.playerCards[winner.id] || [],
                handRank: 'win',
                handDescription: '其他玩家弃牌',
                isWinner: true,
                winAmount: netWin,
            });
            this.state.phase = poker_1.GamePhase.ENDED;
            return { winners, potResults, allHands };
        }
        const playerHands = new Map();
        for (const player of activePlayers) {
            const holeCards = this.state.playerCards[player.id];
            const communityCards = this.state.communityCards;
            if (holeCards && communityCards.length >= 3) {
                const allCards = [...holeCards, ...communityCards];
                const hand = HandEvaluator_1.HandEvaluator.evaluate(allCards);
                playerHands.set(player.id, { hand, cards: allCards });
            }
        }
        const rankNames = {
            [poker_1.HandRank.HIGH_CARD]: '高牌',
            [poker_1.HandRank.ONE_PAIR]: '一对',
            [poker_1.HandRank.TWO_PAIR]: '两对',
            [poker_1.HandRank.THREE_OF_A_KIND]: '三条',
            [poker_1.HandRank.STRAIGHT]: '顺子',
            [poker_1.HandRank.FLUSH]: '同花',
            [poker_1.HandRank.FULL_HOUSE]: '葫芦',
            [poker_1.HandRank.FOUR_OF_A_KIND]: '四条',
            [poker_1.HandRank.STRAIGHT_FLUSH]: '同花顺',
            [poker_1.HandRank.ROYAL_FLUSH]: '皇家同花顺',
        };
        const winnerIdSet = new Set();
        const winnerGrossWin = new Map();
        for (const pot of this.state.pots) {
            const eligiblePlayers = activePlayers.filter(p => pot.eligiblePlayers.includes(p.id));
            if (eligiblePlayers.length === 0)
                continue;
            let bestHand = null;
            let potWinnerIds = [];
            for (const player of eligiblePlayers) {
                const handData = playerHands.get(player.id);
                if (!handData)
                    continue;
                const hand = handData.hand;
                if (!bestHand || hand.rank > bestHand.rank ||
                    (hand.rank === bestHand.rank && hand.value > bestHand.value)) {
                    bestHand = hand;
                    potWinnerIds = [player.id];
                }
                else if (hand.rank === bestHand.rank && hand.value === bestHand.value) {
                    potWinnerIds.push(player.id);
                }
            }
            const splitAmount = Math.floor(pot.amount / potWinnerIds.length);
            const remainder = pot.amount - splitAmount * potWinnerIds.length;
            for (let i = 0; i < potWinnerIds.length; i++) {
                const winnerId = potWinnerIds[i];
                winnerIdSet.add(winnerId);
                const winner = this.players.find(p => p.id === winnerId);
                const hand = playerHands.get(winnerId)?.hand;
                const actualWin = splitAmount + (i === 0 ? remainder : 0);
                winnerGrossWin.set(winnerId, (winnerGrossWin.get(winnerId) || 0) + actualWin);
                if (winner) {
                    winners.push({
                        playerId: winner.id,
                        playerName: winner.name,
                        winAmount: actualWin,
                        potType: pot.id === this.state.pots[0]?.id ? 'main' : 'side',
                        handRank: hand ? rankNames[hand.rank] || '未知' : '未知',
                        handDescription: hand?.description || '未知',
                        winningCards: playerHands.get(winnerId)?.hand.cards || [],
                        holeCards: this.state.playerCards[winnerId] || [],
                        explanation: `${winner.name}以${hand ? rankNames[hand.rank] : '未知'}获胜`,
                    });
                }
            }
            potResults.push({
                potId: pot.id,
                amount: pot.amount,
                winners: potWinnerIds,
                splitAmount,
                remainder,
            });
        }
        const winnerPotTypes = new Map();
        for (const w of winners) {
            if (!winnerPotTypes.has(w.playerId)) {
                winnerPotTypes.set(w.playerId, new Set());
            }
            winnerPotTypes.get(w.playerId).add(w.potType);
        }
        for (const player of activePlayers) {
            const handData = playerHands.get(player.id);
            const hand = handData?.hand;
            const initialChips = this.playerInitialChips.get(player.id) || 0;
            const netWin = player.chips - initialChips;
            const isWinner = netWin > 0;
            let potType;
            if (isWinner && winnerIdSet.has(player.id)) {
                const types = winnerPotTypes.get(player.id);
                if (types) {
                    if (types.has('both')) {
                        potType = 'both';
                    }
                    else if (types.has('side') && types.has('main')) {
                        potType = 'both';
                    }
                    else if (types.has('side')) {
                        potType = 'side';
                    }
                    else {
                        potType = 'main';
                    }
                }
            }
            allHands.push({
                playerId: player.id,
                playerName: player.name,
                holeCards: this.state.playerCards[player.id] || [],
                handRank: hand ? rankNames[hand.rank] || '未知' : '未知',
                handDescription: hand?.description || '未知',
                isWinner,
                winAmount: isWinner ? netWin : undefined,
                potType,
                netWin,
            });
        }
        this.state.phase = poker_1.GamePhase.ENDED;
        return { winners, potResults, allHands };
    }
    getPotAmount() {
        return this.calcTotalPot();
    }
    calcTotalPot() {
        return this.state.pots.reduce((sum, p) => sum + p.amount, 0) +
            Object.values(this.state.roundBets).reduce((sum, b) => sum + (b || 0), 0);
    }
    getState() {
        return { ...this.state };
    }
    getPlayerCards(playerId) {
        return this.state.playerCards[playerId];
    }
    getPlayers() {
        return this.players.map(p => ({ ...p }));
    }
    getCurrentPlayerId() {
        if (this.state.currentPlayerIndex < 0 || this.state.currentPlayerIndex >= this.players.length) {
            return undefined;
        }
        return this.players[this.state.currentPlayerIndex]?.id;
    }
    getValidActions(playerId) {
        const player = this.players.find(p => p.id === playerId);
        if (!player)
            return [];
        const status = this.state.playerStatus[playerId];
        if (status !== poker_1.PlayerStatus.PLAYING)
            return [];
        const myBet = this.state.roundBets[playerId] || 0;
        const toCall = this.state.currentBet - myBet;
        const actions = ['fold'];
        if (toCall <= 0) {
            actions.push('check');
        }
        else if (player.chips >= toCall) {
            actions.push('call');
        }
        if (toCall > 0 && player.chips > 0 && player.chips <= toCall) {
            actions.push('all-in');
        }
        if (player.chips > toCall) {
            actions.push('raise');
        }
        if (player.chips > 0 && player.chips > toCall) {
            actions.push('all-in');
        }
        return actions;
    }
}
exports.GameEngine = GameEngine;
