// texas-poker-server/dist/game/SinglePlayerGameEngine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SinglePlayerGameEngine = void 0;
const uuid_1 = require("uuid");
const poker_1 = require("../types/poker");
const Deck_1 = require("../poker/Deck");
const HandEvaluator_1 = require("../poker/HandEvaluator");
class SinglePlayerGameEngine {
    constructor(config, gameConfig) {
        this.npcTimeouts = [];
        this.isRunning = false;
        this.hasActedThisRound = new Set();
        this.lastAggressorIndex = -1;
        this.actionCount = 0;
        this.playerInitialChips = new Map();
        this.humanPlayerId = config.humanPlayerId;
        this.config = gameConfig;
        this.deck = new Deck_1.Deck();
        this.players = this.createPlayers(config);
        const playerIds = this.players.map(p => p.id);
        this.state = {
            handId: (0, uuid_1.v4)(),
            phase: poker_1.GamePhase.WAITING,
            deck: [],
            communityCards: [],
            pots: [],
            totalPot: 0,
            currentPlayerIndex: -1,
            currentPlayerId: '',
            dealerIndex: 0,
            smallBlindIndex: 1,
            bigBlindIndex: 2,
            lastRaiseIndex: -1,
            currentBet: 0,
            minRaise: gameConfig.bigBlind,
            roundBets: {},
            playerCards: {},
            playerStatus: Object.fromEntries(playerIds.map(id => [id, poker_1.PlayerStatus.WAITING])),
            playerRoles: {},
            actions: [],
            startTime: Date.now(),
        };
    }
    createPlayers(config) {
        const players = [];
        players.push({
            id: config.humanPlayerId,
            name: config.humanPlayerName || '玩家',
            avatar: '',
            seatIndex: 0,
            chips: config.buyIn,
            totalBuyIn: config.buyIn,
            isReady: true,
            isOnline: true,
            isNpc: false,
            joinedAt: Date.now(),
        });
        const npcNames = ['小明', '小红', '小强', '小李', '小张', '小王', '小刘'];
        for (let i = 0; i < config.npcCount; i++) {
            const npcId = `npc_${Date.now()}_${i}`;
            players.push({
                id: npcId,
                name: npcNames[i % npcNames.length],
                avatar: '',
                seatIndex: i + 1,
                chips: config.buyIn,
                totalBuyIn: config.buyIn,
                isReady: true,
                isOnline: true,
                isNpc: true,
                joinedAt: Date.now(),
            });
        }
        return players;
    }
    getState() {
        return { ...this.state };
    }
    getPlayers() {
        return this.players.map(p => ({ ...p }));
    }
    getPlayerCards(playerId) {
        return this.state.playerCards[playerId];
    }
    isHumanPlayerTurn() {
        if (this.state.phase === poker_1.GamePhase.SHOWDOWN || this.state.phase === poker_1.GamePhase.ENDED) {
            return false;
        }
        const currentPlayer = this.players[this.state.currentPlayerIndex];
        return currentPlayer?.id === this.humanPlayerId &&
            this.state.playerStatus[currentPlayer.id] === poker_1.PlayerStatus.PLAYING;
    }
    getPotAmount() {
        return this.state.pots.reduce((sum, p) => sum + p.amount, 0) +
            Object.values(this.state.roundBets).reduce((sum, b) => sum + (b || 0), 0);
    }
    start() {
        this.isRunning = true;
        this.deck = new Deck_1.Deck();
        this.deck.shuffle();
        this.state.communityCards = [];
        this.state.pots = [];
        this.state.roundBets = {};
        this.state.currentBet = 0;
        this.state.lastRaiseIndex = -1;
        this.state.actions = [];
        this.hasActedThisRound = new Set();
        this.lastAggressorIndex = -1;
        this.actionCount = 0;
        for (const player of this.players) {
            this.state.playerStatus[player.id] = poker_1.PlayerStatus.PLAYING;
            this.state.roundBets[player.id] = 0;
            this.playerInitialChips.set(player.id, player.chips);
        }
        this.state.phase = poker_1.GamePhase.PRE_FLOP;
        this.initPlayerRoles();
        this.dealHoleCards();
        this.postBlinds();
        this.state.currentPlayerIndex = this.getNextActivePlayerIndex(this.state.bigBlindIndex);
        this.state.currentPlayerId = this.players[this.state.currentPlayerIndex]?.id || '';
        this.scheduleNpcActions();
    }
    initPlayerRoles() {
        this.state.playerRoles = {};
        const n = this.players.length;
        const dealer = this.players[this.state.dealerIndex % n];
        const sb = this.players[this.state.smallBlindIndex % n];
        const bb = this.players[this.state.bigBlindIndex % n];
        if (dealer)
            this.state.playerRoles[dealer.id] = poker_1.PlayerRole.DEALER;
        if (sb)
            this.state.playerRoles[sb.id] = poker_1.PlayerRole.SB;
        if (bb)
            this.state.playerRoles[bb.id] = poker_1.PlayerRole.BB;
    }
    dealHoleCards() {
        for (const player of this.players) {
            const card1 = this.deck.deal();
            const card2 = this.deck.deal();
            this.state.playerCards[player.id] = [card1, card2];
        }
    }
    postBlinds() {
        const n = this.players.length;
        const sb = this.players[this.state.smallBlindIndex % n];
        const bb = this.players[this.state.bigBlindIndex % n];
        if (sb) {
            const sbAmount = Math.min(this.config.smallBlind, sb.chips);
            sb.chips -= sbAmount;
            this.state.roundBets[sb.id] = sbAmount;
        }
        if (bb) {
            const bbAmount = Math.min(this.config.bigBlind, bb.chips);
            bb.chips -= bbAmount;
            this.state.roundBets[bb.id] = bbAmount;
            this.state.currentBet = bbAmount;
        }
        this.state.minRaise = this.config.bigBlind;
        this.lastAggressorIndex = this.state.bigBlindIndex % n;
    }
    getNextActivePlayerIndex(fromIndex) {
        const n = this.players.length;
        let index = (fromIndex + 1) % n;
        let checks = 0;
        while (checks < n) {
            const player = this.players[index];
            const status = this.state.playerStatus[player.id];
            if (status === poker_1.PlayerStatus.PLAYING) {
                return index;
            }
            index = (index + 1) % n;
            checks++;
        }
        return -1;
    }
    scheduleNpcActions() {
        if (!this.isRunning)
            return;
        const currentPlayer = this.players[this.state.currentPlayerIndex];
        if (!currentPlayer)
            return;
        if (currentPlayer.id === this.humanPlayerId) {
            return;
        }
        if (this.state.playerStatus[currentPlayer.id] !== poker_1.PlayerStatus.PLAYING) {
            this.advanceToNextPlayer();
            return;
        }
        const timeout = setTimeout(() => {
            if (!this.isRunning)
                return;
            const action = this.decideNpcAction(currentPlayer);
            this.executeAction(currentPlayer.id, action.action, action.amount);
        }, 800 + Math.random() * 700);
        this.npcTimeouts.push(timeout);
    }
    decideNpcAction(player) {
        const myBet = this.state.roundBets[player.id] || 0;
        const toCall = this.state.currentBet - myBet;
        const myChips = player.chips;
        const random = Math.random();
        const cards = this.state.playerCards[player.id];
        const hasHighCard = cards && (cards[0].rank === 'A' || cards[0].rank === 'K' || cards[0].rank === 'Q' ||
            cards[1].rank === 'A' || cards[1].rank === 'K' || cards[1].rank === 'Q');
        const isPair = cards && cards[0].rank === cards[1].rank;
        if (toCall === 0) {
            if (isPair && random < 0.6 && myChips > this.config.bigBlind * 3) {
                const raiseAmount = Math.min(this.config.bigBlind * 3, myChips);
                return { action: poker_1.PlayerAction.RAISE, amount: raiseAmount };
            }
            if (hasHighCard && random < 0.3 && myChips > this.config.bigBlind * 2) {
                const raiseAmount = Math.min(this.config.bigBlind * 2, myChips);
                return { action: poker_1.PlayerAction.RAISE, amount: raiseAmount };
            }
            if (random < 0.15 && myChips > this.config.bigBlind * 2) {
                const raiseAmount = Math.min(this.config.bigBlind * 2, myChips);
                return { action: poker_1.PlayerAction.RAISE, amount: raiseAmount };
            }
            return { action: poker_1.PlayerAction.CHECK };
        }
        if (isPair) {
            if (random < 0.7 && myChips >= toCall) {
                return { action: poker_1.PlayerAction.CALL, amount: toCall };
            }
            if (random < 0.9 && myChips > toCall * 2) {
                const raiseAmount = Math.min(toCall * 2, myChips);
                return { action: poker_1.PlayerAction.RAISE, amount: raiseAmount };
            }
        }
        if (hasHighCard) {
            if (toCall <= myChips * 0.3) {
                if (random < 0.6 && myChips >= toCall) {
                    return { action: poker_1.PlayerAction.CALL, amount: toCall };
                }
                if (random < 0.8 && myChips > toCall) {
                    const raiseAmount = Math.min(toCall * 2, myChips);
                    return { action: poker_1.PlayerAction.RAISE, amount: raiseAmount };
                }
            }
            if (random < 0.4 && myChips >= toCall) {
                return { action: poker_1.PlayerAction.CALL, amount: toCall };
            }
        }
        if (toCall > myChips * 0.5) {
            if (random < 0.2 && myChips >= toCall) {
                return { action: poker_1.PlayerAction.CALL, amount: toCall };
            }
            return { action: poker_1.PlayerAction.FOLD };
        }
        if (random < 0.4 && myChips >= toCall) {
            return { action: poker_1.PlayerAction.CALL, amount: toCall };
        }
        return { action: poker_1.PlayerAction.FOLD };
    }
    executeAction(playerId, action, amount) {
        if (!this.isRunning)
            return false;
        const playerIndex = this.players.findIndex(p => p.id === playerId);
        if (playerIndex === -1)
            return false;
        const player = this.players[playerIndex];
        if (this.state.playerStatus[player.id] !== poker_1.PlayerStatus.PLAYING)
            return false;
        if (playerIndex !== this.state.currentPlayerIndex)
            return false;
        const myBet = this.state.roundBets[player.id] || 0;
        const toCall = this.state.currentBet - myBet;
        switch (action) {
            case poker_1.PlayerAction.FOLD:
                this.state.playerStatus[player.id] = poker_1.PlayerStatus.FOLDED;
                break;
            case poker_1.PlayerAction.CHECK:
                if (toCall > 0)
                    return false;
                break;
            case poker_1.PlayerAction.CALL: {
                const callAmount = Math.min(toCall, player.chips);
                if (callAmount <= 0)
                    return false;
                player.chips -= callAmount;
                this.state.roundBets[player.id] = myBet + callAmount;
                if (player.chips === 0) {
                    this.state.playerStatus[player.id] = poker_1.PlayerStatus.ALL_IN;
                }
                break;
            }
            case poker_1.PlayerAction.RAISE: {
                const totalRaiseAmount = amount || this.config.bigBlind * 2;
                const callPlusRaise = toCall + totalRaiseAmount;
                const actualAmount = Math.min(callPlusRaise, player.chips);
                player.chips -= actualAmount;
                this.state.roundBets[player.id] = myBet + actualAmount;
                this.state.currentBet = myBet + actualAmount;
                this.state.minRaise = totalRaiseAmount;
                this.lastAggressorIndex = playerIndex;
                this.hasActedThisRound.add(player.id);
                if (player.chips === 0) {
                    this.state.playerStatus[player.id] = poker_1.PlayerStatus.ALL_IN;
                }
                break;
            }
            case poker_1.PlayerAction.ALL_IN: {
                const allInAmount = player.chips;
                player.chips = 0;
                this.state.roundBets[player.id] = myBet + allInAmount;
                if (myBet + allInAmount > this.state.currentBet) {
                    this.state.currentBet = myBet + allInAmount;
                    this.lastAggressorIndex = playerIndex;
                }
                this.state.playerStatus[player.id] = poker_1.PlayerStatus.ALL_IN;
                this.hasActedThisRound.add(player.id);
                break;
            }
            default:
                return false;
        }
        if (action !== poker_1.PlayerAction.RAISE && action !== poker_1.PlayerAction.ALL_IN) {
            this.hasActedThisRound.add(player.id);
        }
        this.actionCount++;
        this.state.actions.push({
            playerId,
            playerName: player.name,
            action: action.toString(),
            amount: amount || 0,
            timestamp: Date.now(),
            phase: this.state.phase,
        });
        if (this.checkOnlyOnePlayerLeft()) {
            this.endHand();
            return true;
        }
        if (this.isBettingRoundComplete()) {
            this.advancePhase();
        }
        else {
            this.advanceToNextPlayer();
        }
        return true;
    }
    checkOnlyOnePlayerLeft() {
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.PLAYING);
        const allInPlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.ALL_IN);
        return activePlayers.length + allInPlayers.length <= 1;
    }
    isBettingRoundComplete() {
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.PLAYING);
        if (activePlayers.length === 0) {
            return true;
        }
        if (activePlayers.length === 1) {
            const player = activePlayers[0];
            const myBet = this.state.roundBets[player.id] || 0;
            if (myBet >= this.state.currentBet && this.hasActedThisRound.has(player.id)) {
                return true;
            }
        }
        for (const player of activePlayers) {
            const myBet = this.state.roundBets[player.id] || 0;
            if (myBet < this.state.currentBet) {
                return false;
            }
            if (!this.hasActedThisRound.has(player.id)) {
                return false;
            }
        }
        return true;
    }
    advanceToNextPlayer() {
        const nextIndex = this.getNextActivePlayerIndex(this.state.currentPlayerIndex);
        if (nextIndex === -1) {
            this.advancePhase();
            return;
        }
        this.state.currentPlayerIndex = nextIndex;
        this.state.currentPlayerId = this.players[nextIndex]?.id || '';
        this.scheduleNpcActions();
    }
    advancePhase() {
        this.collectBets();
        this.hasActedThisRound = new Set();
        this.lastAggressorIndex = -1;
        this.state.currentBet = 0;
        this.state.roundBets = {};
        this.actionCount = 0;
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.PLAYING ||
            this.state.playerStatus[p.id] === poker_1.PlayerStatus.ALL_IN);
        if (activePlayers.length <= 1) {
            this.endHand();
            return;
        }
        const playingPlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.PLAYING);
        switch (this.state.phase) {
            case poker_1.GamePhase.PRE_FLOP:
                this.state.phase = poker_1.GamePhase.FLOP;
                this.deck.deal();
                this.state.communityCards.push(this.deck.deal(), this.deck.deal(), this.deck.deal());
                break;
            case poker_1.GamePhase.FLOP:
                this.state.phase = poker_1.GamePhase.TURN;
                this.deck.deal();
                this.state.communityCards.push(this.deck.deal());
                break;
            case poker_1.GamePhase.TURN:
                this.state.phase = poker_1.GamePhase.RIVER;
                this.deck.deal();
                this.state.communityCards.push(this.deck.deal());
                break;
            case poker_1.GamePhase.RIVER:
                this.endHand();
                return;
            default:
                return;
        }
        if (playingPlayers.length <= 1) {
            this.dealRemainingCommunityCards();
            this.endHand();
            return;
        }
        this.state.currentPlayerIndex = this.getNextActivePlayerIndex(this.state.dealerIndex);
        this.state.currentPlayerId = this.players[this.state.currentPlayerIndex]?.id || '';
        this.scheduleNpcActions();
    }
    dealRemainingCommunityCards() {
        while (this.state.communityCards.length < 5) {
            this.deck.deal();
            this.state.communityCards.push(this.deck.deal());
        }
        this.state.phase = poker_1.GamePhase.RIVER;
    }
    collectBets() {
        let totalBets = 0;
        for (const player of this.players) {
            const bet = this.state.roundBets[player.id] || 0;
            totalBets += bet;
        }
        if (totalBets > 0) {
            const eligiblePlayers = this.players
                .filter(p => this.state.playerStatus[p.id] !== poker_1.PlayerStatus.FOLDED)
                .map(p => p.id);
            this.state.pots.push({
                id: (0, uuid_1.v4)(),
                amount: totalBets,
                eligiblePlayers,
            });
        }
    }
    endHand() {
        this.collectBets();
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] !== poker_1.PlayerStatus.FOLDED);
        if (activePlayers.length === 0) {
            this.state.phase = poker_1.GamePhase.SHOWDOWN;
            return;
        }
        if (activePlayers.length === 1) {
            const winner = activePlayers[0];
            const totalPot = this.state.pots.reduce((sum, p) => sum + p.amount, 0);
            winner.chips += totalPot;
            this.state.phase = poker_1.GamePhase.SHOWDOWN;
            return;
        }
        this.determineWinners();
        this.state.phase = poker_1.GamePhase.SHOWDOWN;
    }
    determineWinners() {
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] !== poker_1.PlayerStatus.FOLDED);
        if (activePlayers.length === 0)
            return;
        const playerHands = new Map();
        for (const player of activePlayers) {
            const holeCards = this.state.playerCards[player.id];
            const communityCards = this.state.communityCards;
            if (holeCards && communityCards.length >= 3) {
                const allCards = [...holeCards, ...communityCards];
                const hand = HandEvaluator_1.HandEvaluator.evaluate(allCards);
                playerHands.set(player.id, { hand, cards: allCards });
            }
        }
        for (const pot of this.state.pots) {
            const eligiblePlayers = activePlayers.filter(p => pot.eligiblePlayers.includes(p.id));
            if (eligiblePlayers.length === 0)
                continue;
            let bestHand = null;
            let potWinnerIds = [];
            for (const player of eligiblePlayers) {
                const handData = playerHands.get(player.id);
                if (!handData)
                    continue;
                const hand = handData.hand;
                if (!bestHand || hand.rank > bestHand.rank ||
                    (hand.rank === bestHand.rank && hand.value > bestHand.value)) {
                    bestHand = hand;
                    potWinnerIds = [player.id];
                }
                else if (hand.rank === bestHand.rank && hand.value === bestHand.value) {
                    potWinnerIds.push(player.id);
                }
            }
            const splitAmount = Math.floor(pot.amount / potWinnerIds.length);
            const remainder = pot.amount - splitAmount * potWinnerIds.length;
            for (let i = 0; i < potWinnerIds.length; i++) {
                const winner = this.players.find(p => p.id === potWinnerIds[i]);
                if (winner) {
                    winner.chips += splitAmount + (i === 0 ? remainder : 0);
                }
            }
        }
    }
    getWinners() {
        return this.getWinnersAndAllHands().winners;
    }
    getAllHands() {
        return this.getWinnersAndAllHands().allHands;
    }
    getWinnersAndAllHands() {
        const activePlayers = this.players.filter(p => this.state.playerStatus[p.id] !== poker_1.PlayerStatus.FOLDED);
        const allHands = [];
        if (activePlayers.length === 0)
            return { winners: [], allHands };
        const rankNames = {
            [poker_1.HandRank.HIGH_CARD]: '高牌',
            [poker_1.HandRank.ONE_PAIR]: '一对',
            [poker_1.HandRank.TWO_PAIR]: '两对',
            [poker_1.HandRank.THREE_OF_A_KIND]: '三条',
            [poker_1.HandRank.STRAIGHT]: '顺子',
            [poker_1.HandRank.FLUSH]: '同花',
            [poker_1.HandRank.FULL_HOUSE]: '葫芦',
            [poker_1.HandRank.FOUR_OF_A_KIND]: '四条',
            [poker_1.HandRank.STRAIGHT_FLUSH]: '同花顺',
            [poker_1.HandRank.ROYAL_FLUSH]: '皇家同花顺',
        };
        if (activePlayers.length === 1) {
            const winner = activePlayers[0];
            const totalPot = this.state.pots.reduce((sum, p) => sum + p.amount, 0);
            const initialChips = this.playerInitialChips.get(winner.id) || 0;
            const netWin = winner.chips - initialChips;
            allHands.push({
                playerId: winner.id,
                playerName: winner.name,
                holeCards: this.state.playerCards[winner.id] || [],
                handRank: '其他玩家弃牌',
                handDescription: '其他玩家弃牌',
                isWinner: true,
                winAmount: netWin,
            });
            const foldedPlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.FOLDED);
            for (const fp of foldedPlayers) {
                allHands.push({
                    playerId: fp.id,
                    playerName: fp.name,
                    holeCards: this.state.playerCards[fp.id] || [],
                    handRank: '弃牌',
                    handDescription: '弃牌',
                    isWinner: false,
                });
            }
            return {
                winners: [{
                        playerId: winner.id,
                        playerName: winner.name,
                        handRank: 'win',
                        handDescription: '其他玩家弃牌',
                        winningCards: this.state.playerCards[winner.id] || [],
                        holeCards: this.state.playerCards[winner.id] || [],
                        explanation: `${winner.name}获胜，其他玩家弃牌`,
                        winAmount: netWin,
                        potType: 'main',
                    }],
                allHands,
            };
        }
        const playerHands = new Map();
        for (const player of activePlayers) {
            const holeCards = this.state.playerCards[player.id];
            const communityCards = this.state.communityCards;
            if (holeCards && communityCards.length >= 3) {
                const allCards = [...holeCards, ...communityCards];
                const hand = HandEvaluator_1.HandEvaluator.evaluate(allCards);
                playerHands.set(player.id, { hand, cards: allCards });
            }
        }
        const winnerIdSet = new Set();
        const winnerAmounts = new Map();
        for (const pot of this.state.pots) {
            const eligiblePlayers = activePlayers.filter(p => pot.eligiblePlayers.includes(p.id));
            if (eligiblePlayers.length === 0)
                continue;
            let bestHand = null;
            let potWinnerIds = [];
            for (const player of eligiblePlayers) {
                const handData = playerHands.get(player.id);
                if (!handData)
                    continue;
                const hand = handData.hand;
                if (!bestHand || hand.rank > bestHand.rank ||
                    (hand.rank === bestHand.rank && hand.value > bestHand.value)) {
                    bestHand = hand;
                    potWinnerIds = [player.id];
                }
                else if (hand.rank === bestHand.rank && hand.value === bestHand.value) {
                    potWinnerIds.push(player.id);
                }
            }
            const splitAmount = Math.floor(pot.amount / potWinnerIds.length);
            const remainder = pot.amount - splitAmount * potWinnerIds.length;
            for (let i = 0; i < potWinnerIds.length; i++) {
                const wid = potWinnerIds[i];
                winnerIdSet.add(wid);
                const actualWin = splitAmount + (i === 0 ? remainder : 0);
                winnerAmounts.set(wid, (winnerAmounts.get(wid) || 0) + actualWin);
            }
        }
        const winnerPotTypes = new Map();
        for (const pot of this.state.pots) {
            const eligiblePlayers = activePlayers.filter(p => pot.eligiblePlayers.includes(p.id));
            if (eligiblePlayers.length === 0)
                continue;
            let bestHand = null;
            let potWinnerIds = [];
            for (const player of eligiblePlayers) {
                const handData = playerHands.get(player.id);
                if (!handData)
                    continue;
                const hand = handData.hand;
                if (!bestHand || hand.rank > bestHand.rank ||
                    (hand.rank === bestHand.rank && hand.value > bestHand.value)) {
                    bestHand = hand;
                    potWinnerIds = [player.id];
                }
                else if (hand.rank === bestHand.rank && hand.value === bestHand.value) {
                    potWinnerIds.push(player.id);
                }
            }
            const isMainPot = pot.id === this.state.pots[0]?.id;
            for (const wid of potWinnerIds) {
                if (!winnerPotTypes.has(wid)) {
                    winnerPotTypes.set(wid, new Set());
                }
                winnerPotTypes.get(wid).add(isMainPot ? 'main' : 'side');
            }
        }
        for (const player of activePlayers) {
            const handData = playerHands.get(player.id);
            const hand = handData?.hand;
            const initialChips = this.playerInitialChips.get(player.id) || 0;
            const netWin = player.chips - initialChips;
            const isWinner = netWin > 0;
            let potType;
            if (isWinner && winnerIdSet.has(player.id)) {
                const types = winnerPotTypes.get(player.id);
                if (types) {
                    if (types.has('both')) {
                        potType = 'both';
                    }
                    else if (types.has('side') && types.has('main')) {
                        potType = 'both';
                    }
                    else if (types.has('side')) {
                        potType = 'side';
                    }
                    else {
                        potType = 'main';
                    }
                }
            }
            allHands.push({
                playerId: player.id,
                playerName: player.name,
                holeCards: this.state.playerCards[player.id] || [],
                handRank: hand ? (rankNames[hand.rank] || '未知') : '未知',
                handDescription: hand?.description || '未知',
                isWinner,
                winAmount: isWinner ? netWin : undefined,
                potType,
                netWin,
            });
        }
        const foldedPlayers = this.players.filter(p => this.state.playerStatus[p.id] === poker_1.PlayerStatus.FOLDED);
        for (const fp of foldedPlayers) {
            allHands.push({
                playerId: fp.id,
                playerName: fp.name,
                holeCards: this.state.playerCards[fp.id] || [],
                handRank: '弃牌',
                handDescription: '弃牌',
                isWinner: false,
            });
        }
        const winners = [];
        for (const winnerId of winnerIdSet) {
            const winner = this.players.find(p => p.id === winnerId);
            const hand = playerHands.get(winnerId)?.hand;
            const initialChips = this.playerInitialChips.get(winnerId) || 0;
            const netWin = winner ? winner.chips - initialChips : 0;
            if (winner && netWin > 0) {
                winners.push({
                    playerId: winner.id,
                    playerName: winner.name,
                    handRank: hand ? (rankNames[hand.rank] || '未知') : '未知',
                    handDescription: hand?.description || '未知',
                    winningCards: playerHands.get(winnerId)?.hand.cards || [],
                    holeCards: this.state.playerCards[winnerId] || [],
                    explanation: `${winner.name}以${hand ? rankNames[hand.rank] : '未知'}获胜`,
                    winAmount: netWin,
                    potType: 'main',
                });
            }
        }
        return {
            winners,
            allHands,
        };
    }
    rebuy(playerId, amount) {
        const player = this.players.find(p => p.id === playerId);
        if (!player)
            return;
        player.chips += amount;
        player.totalBuyIn += amount;
    }
    nextHand() {
        this.cleanup();
        this.state.handId = (0, uuid_1.v4)();
        this.state.phase = poker_1.GamePhase.WAITING;
        this.state.communityCards = [];
        this.state.pots = [];
        this.state.roundBets = {};
        this.state.currentBet = 0;
        this.state.lastRaiseIndex = -1;
        this.state.actions = [];
        this.hasActedThisRound = new Set();
        this.lastAggressorIndex = -1;
        this.actionCount = 0;
        this.state.dealerIndex = (this.state.dealerIndex + 1) % this.players.length;
        this.state.smallBlindIndex = (this.state.dealerIndex + 1) % this.players.length;
        this.state.bigBlindIndex = (this.state.dealerIndex + 2) % this.players.length;
        this.start();
    }
    cleanup() {
        this.isRunning = false;
        for (const timeout of this.npcTimeouts) {
            clearTimeout(timeout);
        }
        this.npcTimeouts = [];
    }
}
exports.SinglePlayerGameEngine = SinglePlayerGameEngine;
