// texas-poker-server/dist/poker/Deck.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Deck = void 0;
const poker_1 = require("../types/poker");
class Deck {
    constructor() {
        this.cards = [];
        this.reset();
    }
    // 重置并洗牌
    reset() {
        this.cards = [];
        for (const suit of poker_1.SUITS) {
            for (const rank of poker_1.RANKS) {
                this.cards.push({
                    suit,
                    rank,
                    code: `${rank}${suit[0].toUpperCase()}`,
                });
            }
        }
        this.shuffle();
    }
    // 洗牌 (Fisher-Yates算法)
    shuffle() {
        for (let i = this.cards.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [this.cards[i], this.cards[j]] = [this.cards[j], this.cards[i]];
        }
    }
    // 发牌
    deal() {
        const card = this.cards.pop();
        if (!card) {
            throw new Error('Deck is empty');
        }
        return card;
    }
    // 发多张牌
    dealMultiple(count) {
        const cards = [];
        for (let i = 0; i < count; i++) {
            cards.push(this.deal());
        }
        return cards;
    }
    // 查看剩余牌数
    remaining() {
        return this.cards.length;
    }
    // 烧牌 (弃牌)
    burn() {
        this.deal();
    }
}
exports.Deck = Deck;
