// texas-poker-server/dist/websocket/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupWebSocket = setupWebSocket;
const roomHandler_1 = require("./handlers/roomHandler");
const gameHandler_1 = require("./handlers/gameHandler");
const events_1 = require("../types/events");
const gameHandler_2 = require("./handlers/gameHandler");
function sanitizeRoom(room) {
    return {
        config: room.config,
        status: room.status,
        players: room.players.map((p) => ({
            id: p.id,
            name: p.name,
            avatar: p.avatar,
            seatIndex: p.seatIndex,
            chips: p.chips,
            totalBuyIn: p.totalBuyIn,
            isReady: p.isReady,
            isOnline: p.isOnline,
        })),
    };
}
const playerSocketMap = new Map();
function setupWebSocket(io, roomManager) {
    io.on('connection', (socket) => {
        console.log(`Client connected: ${socket.id}`);
        const queryPlayerId = socket.handshake.query.playerId;
        let playerId;
        let isReconnection = false;
        if (queryPlayerId && roomManager.getPlayerRoomId(queryPlayerId)) {
            playerId = queryPlayerId;
            isReconnection = true;
            console.log(`Player reconnecting: ${playerId}`);
        }
        else {
            playerId = `player_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
        }
        socket.data.playerId = playerId;
        playerSocketMap.set(playerId, socket.id);
        socket.emit(events_1.ServerEvents.CONNECTED, {
            playerId,
            message: isReconnection ? '重新连接成功' : '连接成功',
        });
        if (isReconnection) {
            const roomId = roomManager.getPlayerRoomId(playerId);
            if (roomId) {
                const room = roomManager.getRoom(roomId);
                if (room) {
                    socket.join(roomId);
                    socket.data.roomId = roomId;
                    const player = room.players.find(p => p.id === playerId);
                    if (player) {
                        player.isOnline = true;
                    }
                    io.to(roomId).emit(events_1.ServerEvents.PLAYER_JOINED, {
                        player,
                        room: sanitizeRoom(room),
                    });
                    if (room.gameState) {
                        const gameEngine = gameHandler_2.gameEngines.get(roomId);
                        socket.emit(events_1.ServerEvents.GAME_STARTED, {
                            room: sanitizeRoom(room),
                            gameState: {
                                ...room.gameState,
                                playerCards: {},
                            },
                        });
                        if (gameEngine) {
                            const cards = gameEngine.getPlayerCards(playerId);
                            if (cards) {
                                socket.emit(events_1.ServerEvents.DEAL_CARDS, {
                                    handId: room.gameState.handId,
                                    playerId,
                                    cards,
                                });
                            }
                            const currentPlayerId = gameEngine.getCurrentPlayerId();
                            if (currentPlayerId) {
                                const currentPlayer = room.players.find(p => p.id === currentPlayerId);
                                if (currentPlayer) {
                                    socket.emit(events_1.ServerEvents.PLAYER_TURN, {
                                        playerId: currentPlayerId,
                                        playerName: currentPlayer.name,
                                        timeout: 30,
                                        validActions: gameEngine.getValidActions(currentPlayerId),
                                    });
                                }
                                else {
                                    socket.emit(events_1.ServerEvents.PLAYER_TURN, {
                                        playerId: currentPlayerId,
                                        playerName: currentPlayerId,
                                        timeout: 30,
                                        validActions: gameEngine.getValidActions(currentPlayerId),
                                    });
                                }
                            }
                        }
                    }
                }
            }
        }
        (0, roomHandler_1.handleRoomEvents)(socket, io, roomManager);
        (0, gameHandler_1.handleGameEvents)(socket, io, roomManager);
        socket.on('disconnect', () => {
            console.log(`Client disconnected: ${socket.id}`);
            const roomId = roomManager.getPlayerRoomId(playerId);
            if (roomId) {
                const room = roomManager.getRoom(roomId);
                if (room) {
                    const player = room.players.find(p => p.id === playerId);
                    if (player) {
                        player.isOnline = false;
                        io.to(roomId).emit(events_1.ServerEvents.PLAYER_LEFT, {
                            playerId,
                            room: sanitizeRoom(room),
                            isTemporary: true,
                        });
                    }
                }
            }
            playerSocketMap.delete(playerId);
        });
    });
}
