// texas-poker-server/dist/types/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerEvents = exports.ClientEvents = void 0;
// 客户端 → 服务端 事件
var ClientEvents;
(function (ClientEvents) {
    ClientEvents["CREATE_ROOM"] = "room:create";
    ClientEvents["JOIN_ROOM"] = "room:join";
    ClientEvents["LEAVE_ROOM"] = "room:leave";
    ClientEvents["PLAYER_READY"] = "room:ready";
    ClientEvents["START_GAME"] = "room:start";
    ClientEvents["GET_CHIPS"] = "room:get_chips";
    ClientEvents["PLAYER_ACTION"] = "game:action";
    ClientEvents["SEND_CHAT"] = "chat:send";
    ClientEvents["VOTE_LEAVE"] = "room:vote_leave";
    ClientEvents["VOTE_LEAVE_RESPONSE"] = "room:vote_leave_response";
})(ClientEvents || (exports.ClientEvents = ClientEvents = {}));
// 服务端 → 客户端 事件
var ServerEvents;
(function (ServerEvents) {
    ServerEvents["CONNECTED"] = "connection:connected";
    ServerEvents["ERROR"] = "connection:error";
    ServerEvents["ROOM_CREATED"] = "room:created";
    ServerEvents["ROOM_JOINED"] = "room:joined";
    ServerEvents["ROOM_LEFT"] = "room:left";
    ServerEvents["ROOM_UPDATED"] = "room:updated";
    ServerEvents["PLAYER_JOINED"] = "room:player_joined";
    ServerEvents["PLAYER_LEFT"] = "room:player_left";
    ServerEvents["PLAYER_READY_CHANGED"] = "room:player_ready_changed";
    ServerEvents["GAME_STARTED"] = "game:started";
    ServerEvents["GAME_ENDED"] = "game:ended";
    ServerEvents["DEAL_CARDS"] = "game:deal_cards";
    ServerEvents["COMMUNITY_CARDS"] = "game:community_cards";
    ServerEvents["PLAYER_TURN"] = "game:player_turn";
    ServerEvents["ACTION_RESULT"] = "game:action_result";
    ServerEvents["POT_UPDATED"] = "game:pot_updated";
    ServerEvents["SHOWDOWN"] = "game:showdown";
    ServerEvents["HAND_RESULT"] = "game:hand_result";
    ServerEvents["CHAT_MESSAGE"] = "chat:message";
    ServerEvents["SYSTEM_MESSAGE"] = "system:message";
    ServerEvents["CHIPS_RECEIVED"] = "system:chips_received";
    ServerEvents["ERROR_MESSAGE"] = "system:error";
    ServerEvents["VOTE_LEAVE_STARTED"] = "room:vote_leave_started";
    ServerEvents["VOTE_LEAVE_RESPONSE"] = "room:vote_leave_response";
    ServerEvents["VOTE_LEAVE_ENDED"] = "room:vote_leave_ended";
    ServerEvents["ROOM_CLOSED"] = "room:closed";
})(ServerEvents || (exports.ServerEvents = ServerEvents = {}));
