// texas-poker-server/dist/types/poker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlayerRole = exports.PlayerStatus = exports.PlayerAction = exports.GamePhase = exports.HandRankNames = exports.HandRank = exports.RANKS = exports.SUITS = void 0;
exports.SUITS = ['hearts', 'diamonds', 'clubs', 'spades'];
exports.RANKS = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A'];
// 牌型等级
var HandRank;
(function (HandRank) {
    HandRank[HandRank["HIGH_CARD"] = 1] = "HIGH_CARD";
    HandRank[HandRank["ONE_PAIR"] = 2] = "ONE_PAIR";
    HandRank[HandRank["TWO_PAIR"] = 3] = "TWO_PAIR";
    HandRank[HandRank["THREE_OF_A_KIND"] = 4] = "THREE_OF_A_KIND";
    HandRank[HandRank["STRAIGHT"] = 5] = "STRAIGHT";
    HandRank[HandRank["FLUSH"] = 6] = "FLUSH";
    HandRank[HandRank["FULL_HOUSE"] = 7] = "FULL_HOUSE";
    HandRank[HandRank["FOUR_OF_A_KIND"] = 8] = "FOUR_OF_A_KIND";
    HandRank[HandRank["STRAIGHT_FLUSH"] = 9] = "STRAIGHT_FLUSH";
    HandRank[HandRank["ROYAL_FLUSH"] = 10] = "ROYAL_FLUSH";
})(HandRank || (exports.HandRank = HandRank = {}));
// 牌型名称映射
exports.HandRankNames = {
    [HandRank.HIGH_CARD]: '高牌',
    [HandRank.ONE_PAIR]: '一对',
    [HandRank.TWO_PAIR]: '两对',
    [HandRank.THREE_OF_A_KIND]: '三条',
    [HandRank.STRAIGHT]: '顺子',
    [HandRank.FLUSH]: '同花',
    [HandRank.FULL_HOUSE]: '葫芦',
    [HandRank.FOUR_OF_A_KIND]: '四条',
    [HandRank.STRAIGHT_FLUSH]: '同花顺',
    [HandRank.ROYAL_FLUSH]: '皇家同花顺',
};
// 游戏阶段
var GamePhase;
(function (GamePhase) {
    GamePhase["WAITING"] = "waiting";
    GamePhase["PRE_FLOP"] = "pre-flop";
    GamePhase["FLOP"] = "flop";
    GamePhase["TURN"] = "turn";
    GamePhase["RIVER"] = "river";
    GamePhase["SHOWDOWN"] = "showdown";
    GamePhase["ENDED"] = "ended";
})(GamePhase || (exports.GamePhase = GamePhase = {}));
// 玩家动作
var PlayerAction;
(function (PlayerAction) {
    PlayerAction["FOLD"] = "fold";
    PlayerAction["CHECK"] = "check";
    PlayerAction["CALL"] = "call";
    PlayerAction["RAISE"] = "raise";
    PlayerAction["ALL_IN"] = "all-in";
    PlayerAction["SB"] = "sb";
    PlayerAction["BB"] = "bb";
})(PlayerAction || (exports.PlayerAction = PlayerAction = {}));
// 玩家座位状态
var PlayerStatus;
(function (PlayerStatus) {
    PlayerStatus["EMPTY"] = "empty";
    PlayerStatus["WAITING"] = "waiting";
    PlayerStatus["PLAYING"] = "playing";
    PlayerStatus["FOLDED"] = "folded";
    PlayerStatus["ALL_IN"] = "all-in";
    PlayerStatus["AWAY"] = "away";
})(PlayerStatus || (exports.PlayerStatus = PlayerStatus = {}));
// 玩家座位角色
var PlayerRole;
(function (PlayerRole) {
    PlayerRole["DEALER"] = "dealer";
    PlayerRole["SB"] = "sb";
    PlayerRole["BB"] = "bb";
    PlayerRole["NONE"] = "none";
})(PlayerRole || (exports.PlayerRole = PlayerRole = {}));
