// texas-poker-server/dist/websocket/handlers/roomHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleRoomEvents = handleRoomEvents;
const events_1 = require("../../types/events");
const room_1 = require("../../types/room");
const GameEngine_1 = require("../../game/GameEngine");
const gameHandler_1 = require("./gameHandler");
const ActionLogManager_1 = require("../../room/ActionLogManager");
function safeCallback(callback, response) {
    if (typeof callback === 'function') {
        callback(response);
    }
}
function handleRoomEvents(socket, io, roomManager) {
    socket.on(events_1.ClientEvents.CREATE_ROOM, (data, callback) => {
        try {
            const playerId = socket.data.playerId;
            if (!playerId) {
                safeCallback(callback, { success: false, error: '未登录' });
                return;
            }
            const room = roomManager.createRoom(data, playerId);
            socket.join(room.config.roomId);
            const joinResult = roomManager.joinRoom(room.config.roomId, {
                roomId: room.config.roomId,
                playerName: data.hostName || 'Player',
            }, playerId);
            if (joinResult.success && joinResult.room) {
                socket.emit(events_1.ServerEvents.ROOM_JOINED, {
                    room: sanitizeRoom(joinResult.room),
                    playerId,
                });
                io.to(room.config.roomId).emit(events_1.ServerEvents.PLAYER_JOINED, {
                    player: joinResult.room.players.find(p => p.id === playerId),
                    room: sanitizeRoom(joinResult.room),
                });
                io.emit(events_1.ServerEvents.ROOM_UPDATED, {
                    type: 'created',
                    room: sanitizeRoom(joinResult.room),
                });
                safeCallback(callback, { success: true, room: sanitizeRoom(joinResult.room), playerId });
            }
            else {
                safeCallback(callback, { success: false, error: joinResult.error });
            }
        }
        catch (error) {
            safeCallback(callback, { success: false, error: '创建房间失败' });
        }
    });
    socket.on(events_1.ClientEvents.JOIN_ROOM, (data, callback) => {
        try {
            const playerId = socket.data.playerId;
            if (!playerId) {
                safeCallback(callback, { success: false, error: '未登录' });
                return;
            }
            if (!data.roomId) {
                safeCallback(callback, { success: false, error: '房间ID不能为空' });
                return;
            }
            const result = roomManager.joinRoom(data.roomId, data, playerId);
            if (result.success && result.room) {
                socket.join(data.roomId);
                socket.data.roomId = data.roomId;
                socket.emit(events_1.ServerEvents.ROOM_JOINED, {
                    room: sanitizeRoom(result.room),
                    playerId,
                });
                socket.to(data.roomId).emit(events_1.ServerEvents.PLAYER_JOINED, {
                    player: result.room.players.find(p => p.id === playerId),
                    room: sanitizeRoom(result.room),
                });
                io.emit(events_1.ServerEvents.ROOM_UPDATED, {
                    type: 'updated',
                    room: sanitizeRoom(result.room),
                });
                safeCallback(callback, { success: true, room: sanitizeRoom(result.room), playerId });
            }
            else if (result.error === '你已在该房间中') {
                const room = roomManager.getRoom(data.roomId);
                if (room) {
                    const existingPlayer = room.players.find(p => p.id === playerId);
                    if (existingPlayer && !existingPlayer.isOnline) {
                        existingPlayer.isOnline = true;
                        existingPlayer.name = data.playerName || existingPlayer.name;
                        socket.join(data.roomId);
                        socket.data.roomId = data.roomId;
                        socket.emit(events_1.ServerEvents.ROOM_JOINED, {
                            room: sanitizeRoom(room),
                            playerId,
                        });
                        io.to(data.roomId).emit(events_1.ServerEvents.PLAYER_JOINED, {
                            player: existingPlayer,
                            room: sanitizeRoom(room),
                        });
                        io.emit(events_1.ServerEvents.ROOM_UPDATED, {
                            type: 'updated',
                            room: sanitizeRoom(room),
                        });
                        safeCallback(callback, { success: true, room: sanitizeRoom(room), playerId });
                    }
                    else {
                        safeCallback(callback, { success: false, error: result.error });
                    }
                }
                else {
                    safeCallback(callback, { success: false, error: result.error });
                }
            }
            else {
                safeCallback(callback, { success: false, error: result.error });
            }
        }
        catch (error) {
            safeCallback(callback, { success: false, error: '加入房间失败' });
        }
    });
    socket.on(events_1.ClientEvents.LEAVE_ROOM, (callback) => {
        try {
            const playerId = socket.data.playerId;
            if (!playerId) {
                safeCallback(callback, { success: false, error: '未登录' });
                return;
            }
            const roomId = roomManager.getPlayerRoomId(playerId);
            const result = roomManager.leaveRoom(playerId);
            if (result.success) {
                if (roomId) {
                    socket.leave(roomId);
                    if (result.shouldClose) {
                        io.to(roomId).emit(events_1.ServerEvents.ROOM_CLOSED, {
                            reason: '玩家不足3人，房间已关闭',
                        });
                        (0, ActionLogManager_1.cleanupRoomLogs)(roomId);
                        gameHandler_1.gameEngines.delete(roomId);
                        roomManager.deleteRoom(roomId);
                    }
                    else {
                        const room = roomManager.getRoom(roomId);
                        if (room) {
                            io.to(roomId).emit(events_1.ServerEvents.PLAYER_LEFT, {
                                playerId,
                                room: sanitizeRoom(room),
                            });
                        }
                    }
                }
                socket.data.roomId = null;
                safeCallback(callback, { success: true, shouldClose: result.shouldClose });
            }
            else {
                safeCallback(callback, { success: false, error: result.error });
            }
        }
        catch (error) {
            safeCallback(callback, { success: false, error: '离开房间失败' });
        }
    });
    socket.on(events_1.ClientEvents.PLAYER_READY, (ready, callback) => {
        try {
            const playerId = socket.data.playerId;
            if (!playerId) {
                safeCallback(callback, { success: false, error: '未登录' });
                return;
            }
            const roomId = roomManager.getPlayerRoomId(playerId);
            if (roomId) {
                const room = roomManager.getRoom(roomId);
                if (room && room.status === room_1.RoomStatus.PLAYING) {
                    const gameEngine = gameHandler_1.gameEngines.get(roomId);
                    const isInCurrentGame = gameEngine?.getPlayers().some(p => p.id === playerId);
                    if (isInCurrentGame) {
                        safeCallback(callback, { success: false, error: '你正在游戏中，无需准备' });
                        return;
                    }
                }
            }
            const result = roomManager.setPlayerReady(playerId, ready);
            if (result.success) {
                if (roomId) {
                    const room = roomManager.getRoom(roomId);
                    io.to(roomId).emit(events_1.ServerEvents.PLAYER_READY_CHANGED, {
                        playerId,
                        ready,
                        room: sanitizeRoom(room),
                    });
                }
                safeCallback(callback, { success: true });
            }
            else {
                safeCallback(callback, { success: false, error: result.error });
            }
        }
        catch (error) {
            safeCallback(callback, { success: false, error: '设置准备状态失败' });
        }
    });
    socket.on(events_1.ClientEvents.START_GAME, (callback) => {
        try {
            const playerId = socket.data.playerId;
            if (!playerId) {
                safeCallback(callback, { success: false, error: '未登录' });
                return;
            }
            const roomId = roomManager.getPlayerRoomId(playerId);
            if (!roomId) {
                safeCallback(callback, { success: false, error: '你不在任何房间中' });
                return;
            }
            const room = roomManager.getRoom(roomId);
            if (!room) {
                safeCallback(callback, { success: false, error: '房间不存在' });
                return;
            }
            if (room.config.hostId !== playerId) {
                safeCallback(callback, { success: false, error: '只有庄家才能开始游戏' });
                return;
            }
            let gameEngine = gameHandler_1.gameEngines.get(roomId);
            if (gameEngine && room.status === room_1.RoomStatus.PLAYING) {
                safeCallback(callback, { success: false, error: '游戏正在进行中' });
                return;
            }
            const readyPlayers = room.players.filter(p => p.isReady && p.chips > 0);
            if (readyPlayers.length < 3) {
                safeCallback(callback, { success: false, error: `至少需要3名有筹码的玩家才能开始（当前${readyPlayers.length}人准备且有筹码）` });
                return;
            }
            const gameConfig = {
                smallBlind: room.config.smallBlind,
                bigBlind: room.config.bigBlind,
                actionTimeout: room.config.actionTimeout,
            };
            const dealerIndex = room.gameState ? (room.gameState.dealerIndex + 1) % readyPlayers.length : 0;
            gameEngine = new GameEngine_1.GameEngine(readyPlayers, dealerIndex, gameConfig);
            room.status = room_1.RoomStatus.PLAYING;
            room.gameState = gameEngine.start();
            syncPlayerChipsToRoom(gameEngine, room);
            gameHandler_1.gameEngines.set(roomId, gameEngine);
            io.to(roomId).emit(events_1.ServerEvents.GAME_STARTED, {
                room: sanitizeRoom(room),
                gameState: sanitizeGameState(room.gameState),
            });
            const handId = room.gameState.handId;
            setTimeout(() => {
                for (const player of room.players) {
                    const cards = gameEngine.getPlayerCards(player.id);
                    if (cards) {
                        const playerSockets = Array.from(io.sockets.sockets.values()).filter(s => s.data.playerId === player.id);
                        for (const s of playerSockets) {
                            s.emit(events_1.ServerEvents.DEAL_CARDS, {
                                handId,
                                playerId: player.id,
                                cards,
                            });
                        }
                    }
                }
            }, 100);
            const currentPlayerId = gameEngine.getCurrentPlayerId();
            if (currentPlayerId) {
                const currentPlayer = room.players.find(p => p.id === currentPlayerId);
                if (currentPlayer) {
                    io.to(roomId).emit(events_1.ServerEvents.PLAYER_TURN, {
                        playerId: currentPlayerId,
                        playerName: currentPlayer.name,
                        timeout: 30,
                        validActions: gameEngine.getValidActions(currentPlayerId),
                    });
                }
                else {
                    io.to(roomId).emit(events_1.ServerEvents.PLAYER_TURN, {
                        playerId: currentPlayerId,
                        playerName: currentPlayerId,
                        timeout: 30,
                        validActions: gameEngine.getValidActions(currentPlayerId),
                    });
                }
            }
            safeCallback(callback, { success: true });
        }
        catch (error) {
            safeCallback(callback, { success: false, error: '开始游戏失败' });
        }
    });
    socket.on(events_1.ClientEvents.GET_CHIPS, (callback) => {
        try {
            const playerId = socket.data.playerId;
            if (!playerId) {
                safeCallback(callback, { success: false, error: '未登录' });
                return;
            }
            const result = roomManager.replenishChips(playerId);
            if (result.success) {
                const roomId = roomManager.getPlayerRoomId(playerId);
                if (roomId) {
                    const room = roomManager.getRoom(roomId);
                    const gameEngine = gameHandler_1.gameEngines.get(roomId);
                    if (gameEngine && room) {
                        const enginePlayers = gameEngine.getPlayers();
                        const ep = enginePlayers.find(p => p.id === playerId);
                        if (ep) {
                            ep.chips += (result.amount || 0);
                            syncPlayerChipsToRoom(gameEngine, room);
                        }
                    }
                    io.to(roomId).emit(events_1.ServerEvents.CHIPS_RECEIVED, {
                        playerId,
                        amount: result.amount,
                        room: sanitizeRoom(room),
                    });
                }
                safeCallback(callback, { success: true, amount: result.amount });
            }
            else {
                safeCallback(callback, { success: false, error: result.error });
            }
        }
        catch (error) {
            safeCallback(callback, { success: false, error: '补充筹码失败' });
        }
    });
    socket.on(events_1.ClientEvents.VOTE_LEAVE, (callback) => {
        try {
            const playerId = socket.data.playerId;
            if (!playerId) {
                safeCallback(callback, { success: false, error: '未登录' });
                return;
            }
            const result = roomManager.startVoteLeave(playerId);
            if (result.success) {
                const roomId = roomManager.getPlayerRoomId(playerId);
                if (roomId && result.room) {
                    io.to(roomId).emit(events_1.ServerEvents.VOTE_LEAVE_STARTED, {
                        initiatorId: result.room.voteLeave?.initiatorId,
                        initiatorName: result.room.voteLeave?.initiatorName,
                        votes: Object.fromEntries(result.room.voteLeave?.votes || new Map()),
                        totalPlayers: result.room.players.length,
                    });
                }
                safeCallback(callback, { success: true });
            }
            else {
                safeCallback(callback, { success: false, error: result.error });
            }
        }
        catch (error) {
            safeCallback(callback, { success: false, error: '发起投票失败' });
        }
    });
    socket.on(events_1.ClientEvents.VOTE_LEAVE_RESPONSE, (data, callback) => {
        try {
            const playerId = socket.data.playerId;
            if (!playerId) {
                safeCallback(callback, { success: false, error: '未登录' });
                return;
            }
            const result = roomManager.voteLeaveResponse(playerId, data.approve);
            if (result.success) {
                const roomId = result.roomId || roomManager.getPlayerRoomId(playerId);
                if (roomId && result.room) {
                    io.to(roomId).emit(events_1.ServerEvents.VOTE_LEAVE_RESPONSE, {
                        playerId,
                        approve: data.approve,
                        votes: Object.fromEntries(result.room.voteLeave?.votes || new Map()),
                        totalPlayers: result.room.players.length,
                        votedPlayers: result.room.voteLeave?.votes?.size || 0,
                    });
                    if (result.room.voteLeave && result.room.players.every(p => result.room.voteLeave.votes.has(p.id))) {
                        io.to(roomId).emit(events_1.ServerEvents.VOTE_LEAVE_ENDED, {
                            approved: result.approved,
                            approvedCount: Array.from(result.room.voteLeave.votes.values()).filter(v => v).length,
                            totalPlayers: result.room.players.length,
                        });
                        if (result.approved) {
                            gameHandler_1.gameEngines.delete(roomId);
                            (0, ActionLogManager_1.cleanupRoomLogs)(roomId);
                            io.to(roomId).emit(events_1.ServerEvents.ROOM_LEFT, {
                                reason: 'vote',
                            });
                        }
                    }
                    if (result.approved === false && result.voteCounts) {
                        io.to(roomId).emit(events_1.ServerEvents.VOTE_LEAVE_ENDED, {
                            approved: false,
                            approvedCount: result.voteCounts.approveCount,
                            totalPlayers: result.room.players.length,
                        });
                    }
                }
                if (result.approved && !result.room) {
                    safeCallback(callback, { success: true, approved: true });
                    return;
                }
                safeCallback(callback, { success: true, approved: result.approved });
            }
            else {
                safeCallback(callback, { success: false, error: result.error });
            }
        }
        catch (error) {
            safeCallback(callback, { success: false, error: '响应投票失败' });
        }
    });
}
function syncPlayerChipsToRoom(gameEngine, room) {
    const enginePlayers = gameEngine.getPlayers();
    for (const ep of enginePlayers) {
        const roomPlayer = room.players.find((p) => p.id === ep.id);
        if (roomPlayer) {
            roomPlayer.chips = ep.chips;
        }
    }
}
function sanitizeGameState(gameState) {
    const sanitized = JSON.parse(JSON.stringify(gameState));
    sanitized.playerCards = {};
    return sanitized;
}
function sanitizeRoom(room) {
    return {
        config: room.config,
        status: room.status,
        players: room.players.map((p) => ({
            id: p.id,
            name: p.name,
            avatar: p.avatar,
            seatIndex: p.seatIndex,
            chips: p.chips,
            totalBuyIn: p.totalBuyIn,
            isReady: p.isReady,
            isOnline: p.isOnline,
        })),
    };
}
