// texas-poker-server/src/types/room.ts
import { GamePhase, PlayerStatus, PlayerRole, Card } from './poker';

export interface CreateRoomRequest {
  roomName?: string;
  maxPlayers?: number;
  smallBlind?: number;
  bigBlind?: number;
  buyInMin?: number;
  buyInMax?: number;
  isPrivate?: boolean;
  password?: string;
  hostName?: string;
}

export interface JoinRoomRequest {
  roomId?: string;
  playerName: string;
  password?: string;
}

export interface RoomConfig {
  roomId: string;
  roomName: string;
  hostId: string;
  createdAt: number;
  maxPlayers: number;
  minPlayers: number;
  smallBlind: number;
  bigBlind: number;
  buyInMin: number;
  buyInMax: number;
  actionTimeout: number;
  autoStart: boolean;
  autoStartDelay: number;
  isPrivate: boolean;
  password?: string;
  allowSpectate: boolean;
  allowChat: boolean;
}

export enum RoomStatus {
  WAITING = 'waiting',
  PLAYING = 'playing',
  ENDED = 'ended',
}

export interface RoomPlayer {
  id: string;
  name: string;
  avatar: string;
  seatIndex: number;
  chips: number;
  totalBuyIn: number;
  isReady: boolean;
  isOnline: boolean;
  isNpc?: boolean;
  joinedAt: number;
}

export interface Room {
  config: RoomConfig;
  status: RoomStatus;
  players: RoomPlayer[];
  gameState?: GameState;
  spectators: string[];
  voteLeave?: {
    initiatorId: string;
    initiatorName: string;
    votes: Map<string, boolean>;
    approved: boolean;
  };
}

export interface GameState {
  handId: string;
  phase: GamePhase;
  deck: Card[];
  communityCards: Card[];
  pots: Pot[];
  totalPot: number;
  currentPlayerIndex: number;
  currentPlayerId: string;
  dealerIndex: number;
  smallBlindIndex: number;
  bigBlindIndex: number;
  lastRaiseIndex: number;
  currentBet: number;
  minRaise: number;
  roundBets: Record<string, number>;
  playerCards: Record<string, [Card, Card]>;
  playerStatus: Record<string, PlayerStatus>;
  playerRoles: Record<string, PlayerRole>;
  actions: PlayerActionRecord[];
  startTime: number;
}

export interface Pot {
  id: string;
  amount: number;
  eligiblePlayers: string[];
}

export interface PlayerActionRecord {
  playerId: string;
  playerName: string;
  action: string;
  amount?: number;
  timestamp: number;
  phase: GamePhase;
}

export interface WinnerInfo {
  playerId: string;
  playerName: string;
  winAmount: number;
  potType: 'main' | 'side';
  handRank: string;
  handDescription: string;
  winningCards: Card[];
  holeCards: Card[];
  explanation: string;
}

export interface PlayerHandInfo {
  playerId: string;
  playerName: string;
  holeCards: Card[];
  handRank: string;
  handDescription: string;
  isWinner: boolean;
  winAmount?: number;
  potType?: 'main' | 'side' | 'both';
  netWin?: number;
}

export interface PotResult {
  potId: string;
  amount: number;
  winners: string[];
  splitAmount: number;
  remainder: number;
}
