// texas-poker-server/dist/room/ActionLogManager.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.addActionLog = addActionLog;
exports.getRoomLogs = getRoomLogs;
exports.getHandLogs = getHandLogs;
exports.clearHandLogs = clearHandLogs;
exports.cleanupRoomLogs = cleanupRoomLogs;
exports.loadRoomLogs = loadRoomLogs;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const LOGS_DIR = path.join(process.cwd(), 'temp_logs');
if (!fs.existsSync(LOGS_DIR)) {
    fs.mkdirSync(LOGS_DIR, { recursive: true });
}
const roomLogs = new Map();
function addActionLog(roomId, handId, playerId, playerName, action, amount, phase) {
    const record = {
        id: `${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
        roomId,
        handId,
        playerId,
        playerName,
        action,
        amount,
        phase: phase || '',
        timestamp: Date.now(),
    };
    if (!roomLogs.has(roomId)) {
        roomLogs.set(roomId, []);
    }
    roomLogs.get(roomId).push(record);
    persistLogs(roomId);
}
function getRoomLogs(roomId) {
    return roomLogs.get(roomId) || [];
}
function getHandLogs(roomId, handId) {
    const logs = roomLogs.get(roomId) || [];
    return logs.filter(l => l.handId === handId);
}
function clearHandLogs(roomId, handId) {
    const logs = roomLogs.get(roomId);
    if (logs) {
        roomLogs.set(roomId, logs.filter(l => l.handId !== handId));
        persistLogs(roomId);
    }
}
function cleanupRoomLogs(roomId) {
    roomLogs.delete(roomId);
    const filePath = path.join(LOGS_DIR, `${roomId}.json`);
    try {
        if (fs.existsSync(filePath)) {
            fs.unlinkSync(filePath);
        }
    }
    catch { }
}
function persistLogs(roomId) {
    const logs = roomLogs.get(roomId);
    if (!logs)
        return;
    const filePath = path.join(LOGS_DIR, `${roomId}.json`);
    try {
        fs.writeFileSync(filePath, JSON.stringify(logs, null, 2), 'utf-8');
    }
    catch { }
}
function loadRoomLogs(roomId) {
    if (roomLogs.has(roomId))
        return;
    const filePath = path.join(LOGS_DIR, `${roomId}.json`);
    try {
        if (fs.existsSync(filePath)) {
            const data = fs.readFileSync(filePath, 'utf-8');
            roomLogs.set(roomId, JSON.parse(data));
        }
    }
    catch { }
}
