// texas-poker-client/src/stores/socketStore.ts
import { create } from 'zustand'
import { io, Socket } from 'socket.io-client'
import { ServerEvents } from '../types'

// @ts-ignore
const SOCKET_URL = import.meta.env?.VITE_WS_URL || window.location.origin

interface SocketState {
  socket: Socket | null
  isConnected: boolean
  isReconnecting: boolean
  playerId: string | null
  connect: () => void
  disconnect: () => void
  emit: (event: string, data?: any, timeoutMs?: number) => Promise<any>
  on: (event: string, callback: (data: any) => void) => void
  off: (event: string, callback?: (data: any) => void) => void
}

const registeredListeners: Array<{ event: string; callback: (data: any) => void }> = []
const PLAYER_ID_KEY = 'poker_player_id'

function savePlayerId(id: string) {
  try {
    sessionStorage.setItem(PLAYER_ID_KEY, id)
  } catch {}
}

function loadPlayerId(): string | null {
  try {
    return sessionStorage.getItem(PLAYER_ID_KEY)
  } catch {
    return null
  }
}


function rebindListeners(socket: Socket) {
  for (const { event, callback } of registeredListeners) {
    socket.off(event, callback)
    socket.on(event, callback)
  }
}

export const useSocketStore = create<SocketState>((set, get) => ({
  socket: null,
  isConnected: false,
  isReconnecting: false,
  playerId: null,

  connect: () => {
    const savedPlayerId = loadPlayerId()

    const socket = io(SOCKET_URL, {
      transports: ['websocket', 'polling'],
      reconnection: true,
      reconnectionAttempts: Infinity,
      reconnectionDelay: 1000,
      reconnectionDelayMax: 5000,
      timeout: 20000,
      query: savedPlayerId ? { playerId: savedPlayerId } : {},
    })

    socket.on('connect', () => {
      console.log('Socket connected:', socket.id)
      set({ isConnected: true, isReconnecting: false })
      rebindListeners(socket)
    })

    socket.on('disconnect', (reason) => {
      console.log('Socket disconnected, reason:', reason)
      set({ isConnected: false })
      if (reason === 'io server disconnect') {
        socket.connect()
      }
    })

    socket.on('connect_error', (error) => {
      console.error('Socket connect error:', error.message)
      set({ isReconnecting: true })
    })

    socket.on('reconnect_attempt', (attempt) => {
      console.log('Reconnection attempt:', attempt)
      set({ isReconnecting: true })
    })

    socket.on('reconnect', (attempt) => {
      console.log('Reconnected after', attempt, 'attempts')
      set({ isReconnecting: false, isConnected: true })
    })

    socket.on('reconnect_failed', () => {
      console.error('Reconnection failed')
      set({ isReconnecting: false })
    })

    socket.on(ServerEvents.CONNECTED, (data) => {
      const pid = data.playerId || savedPlayerId
      set({ playerId: pid })
      savePlayerId(pid)
      console.log('Connected with playerId:', pid)
    })

    socket.on('error', (error) => {
      console.error('Socket error:', error)
    })

    set({ socket })
  },

  disconnect: () => {
    const { socket } = get()
    if (socket) {
      socket.disconnect()
      set({ socket: null, isConnected: false, isReconnecting: false })
    }
  },

  emit: (event: string, data?: any, timeoutMs?: number) => {
    return new Promise((resolve, reject) => {
      const { socket } = get()
      if (!socket || !socket.connected) {
        reject(new Error('Socket not connected'))
        return
      }

      const timeout = setTimeout(() => {
        reject(new Error('Request timeout'))
      }, timeoutMs || 180000)

      socket.emit(event, data, (response: any) => {
        clearTimeout(timeout)
        if (response?.success) {
          resolve(response)
        } else {
          reject(new Error(response?.error || 'Request failed'))
        }
      })
    })
  },

  on: (event: string, callback: (data: any) => void) => {
    const existing = registeredListeners.find(l => l.event === event && l.callback === callback)
    if (!existing) {
      registeredListeners.push({ event, callback })
    }

    const { socket, isConnected } = get()
    if (socket && isConnected) {
      socket.on(event, callback)
    }
  },

  off: (event: string, callback?: (data: any) => void) => {
    const { socket } = get()
    if (socket) {
      socket.off(event, callback)
    }

    if (callback) {
      const index = registeredListeners.findIndex(
        (l) => l.event === event && l.callback === callback
      )
      if (index !== -1) {
        registeredListeners.splice(index, 1)
      }
    } else {
      for (let i = registeredListeners.length - 1; i >= 0; i--) {
        if (registeredListeners[i].event === event) {
          registeredListeners.splice(i, 1)
        }
      }
    }
  },
}))
