// texas-poker-server/dist/routes/singlePlayerRoutes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const express_1 = require("express");
const SinglePlayerGameEngine_1 = require("../game/SinglePlayerGameEngine");
const poker_1 = require("../types/poker");
const router = (0, express_1.Router)();
const games = new Map();
const defaultGameConfig = {
    smallBlind: 10,
    bigBlind: 20,
    actionTimeout: 30,
};
function buildGameState(game, playerId) {
    const state = game.getState();
    const humanCards = game.getPlayerCards(playerId);
    const isMyTurn = game.isHumanPlayerTurn();
    const winners = state.phase === 'showdown' ? game.getWinners() : [];
    const allHands = state.phase === 'showdown' ? game.getAllHands() : [];
    return {
        gameState: {
            players: game.getPlayers(),
            communityCards: state.communityCards,
            currentPlayerIndex: state.currentPlayerIndex,
            currentPlayerId: state.currentPlayerId,
            phase: state.phase,
            pot: game.getPotAmount(),
            currentBet: state.currentBet,
            minRaise: state.minRaise,
            playerStatus: state.playerStatus,
            playerRoles: state.playerRoles,
            roundBets: state.roundBets,
        },
        humanCards: humanCards || null,
        isMyTurn,
        winners,
        allHands,
    };
}
router.post('/start', (req, res) => {
    try {
        const { playerId, playerName, npcCount, buyIn } = req.body;
        if (!playerId) {
            res.json({ success: false, error: '缺少玩家ID' });
            return;
        }
        const existingGame = games.get(playerId);
        if (existingGame) {
            existingGame.cleanup();
            games.delete(playerId);
        }
        const config = {
            humanPlayerId: playerId,
            humanPlayerName: playerName || '玩家',
            npcCount: npcCount || 3,
            buyIn: buyIn || 1000,
        };
        const game = new SinglePlayerGameEngine_1.SinglePlayerGameEngine(config, defaultGameConfig);
        game.start();
        games.set(playerId, game);
        res.json({
            success: true,
            ...buildGameState(game, playerId),
        });
    }
    catch (error) {
        console.error('Single player start error:', error);
        res.json({ success: false, error: error.message });
    }
});
router.get('/state', (req, res) => {
    try {
        const { playerId } = req.query;
        if (!playerId || typeof playerId !== 'string') {
            res.json({ success: false, error: '缺少玩家ID' });
            return;
        }
        const game = games.get(playerId);
        if (!game) {
            res.json({ success: false, error: '游戏不存在' });
            return;
        }
        res.json({
            success: true,
            ...buildGameState(game, playerId),
        });
    }
    catch (error) {
        console.error('Single player state error:', error);
        res.json({ success: false, error: error.message });
    }
});
router.post('/action', (req, res) => {
    try {
        const { playerId, action, amount } = req.body;
        if (!playerId || !action) {
            res.json({ success: false, error: '缺少参数' });
            return;
        }
        const game = games.get(playerId);
        if (!game) {
            res.json({ success: false, error: '游戏不存在' });
            return;
        }
        const actionMap = {
            'fold': poker_1.PlayerAction.FOLD,
            'check': poker_1.PlayerAction.CHECK,
            'call': poker_1.PlayerAction.CALL,
            'raise': poker_1.PlayerAction.RAISE,
            'allin': poker_1.PlayerAction.ALL_IN,
        };
        const playerAction = actionMap[action.toLowerCase()];
        if (!playerAction) {
            res.json({ success: false, error: `无效操作: ${action}` });
            return;
        }
        const result = game.executeAction(playerId, playerAction, amount);
        if (!result) {
            res.json({ success: false, error: '操作失败' });
            return;
        }
        res.json({
            success: true,
            ...buildGameState(game, playerId),
        });
    }
    catch (error) {
        console.error('Single player action error:', error);
        res.json({ success: false, error: error.message });
    }
});
router.post('/next', (req, res) => {
    try {
        const { playerId } = req.body;
        if (!playerId) {
            res.json({ success: false, error: '缺少玩家ID' });
            return;
        }
        const game = games.get(playerId);
        if (!game) {
            res.json({ success: false, error: '游戏不存在' });
            return;
        }
        game.nextHand();
        res.json({
            success: true,
            ...buildGameState(game, playerId),
        });
    }
    catch (error) {
        console.error('Single player next error:', error);
        res.json({ success: false, error: error.message });
    }
});
router.post('/rebuy', (req, res) => {
    try {
        const { playerId, amount } = req.body;
        if (!playerId) {
            res.json({ success: false, error: '缺少玩家ID' });
            return;
        }
        const game = games.get(playerId);
        if (!game) {
            res.json({ success: false, error: '游戏不存在' });
            return;
        }
        const rebuyAmount = amount || 1000;
        const humanPlayer = game.getPlayers().find(p => p.id === playerId);
        if (!humanPlayer) {
            res.json({ success: false, error: '玩家不存在' });
            return;
        }
        game.rebuy(playerId, rebuyAmount);
        res.json({
            success: true,
            ...buildGameState(game, playerId),
        });
    }
    catch (error) {
        console.error('Single player rebuy error:', error);
        res.json({ success: false, error: error.message });
    }
});
router.post('/end', (req, res) => {
    try {
        const { playerId } = req.body;
        if (!playerId) {
            res.json({ success: false, error: '缺少玩家ID' });
            return;
        }
        const game = games.get(playerId);
        if (game) {
            game.cleanup();
            games.delete(playerId);
        }
        res.json({ success: true });
    }
    catch (error) {
        console.error('Single player end error:', error);
        res.json({ success: false, error: error.message });
    }
});
exports.default = router;
